---
title: "Honeybee cross-species analysis - 09"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---

# Initialization

To intialize complete all of or relevant sections of `scripts/visualization/Load_data.R`.

Due to large differences in total reads, plotting number of filtered pairs directly does not work.
Plotting percentage after normalizing across magOTUs using the factor `length*breadth` is more meaningful


```{r}
system("mkdir -p results/figures/10-instrain_plots/")
```

```{r instrain_scaff_detected}
# plot number of true scaffolds vs detected scaffolds for each sample each magOTU
# showing breadth and coverage...

plot_scaff_detected_list <- list()
for (magotu in sort(unique(instrain_genomes_info_df$magOTU))) {
  plot_scaff_detected <- ggplot(instrain_genomes_info_df_all %>% 
                                mutate(perc_detected = detected_scaffolds/true_scaffolds*100) %>%
                                mutate(detected = ifelse(coverage > 1 & breadth > breadth_cutoff, "y", "n")) %>%
                                mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100) %>%
                                mutate(Genus = as.character(Genus)) %>%
                                mutate(Genus = ifelse(detected == "y", Genus, "g__")) %>%
                                mutate(Host = ifelse(detected == "y", Host, "g__")) %>%
                                filter(magOTU == magotu),
                             aes(y = perc_detected,
                                 x = breadth,
                                 color = Host,
                                 size = coverage_median
                              )
                            ) +
  geom_point() +
  scale_size_continuous(trans = scales::pseudo_log_trans(base = 10)) +
  labs(y = "Percentage of scaffolds detected (at least a read mapped)") +
  ggtitle(paste0(magotu)) +
  # scale_size_continuous(trans = "log10") +
    # facet_wrap_paginate(~magOTU, ncol = 3, nrow = 3, page = 6) +
    geom_vline(xintercept = 50, linetype = "dotted") +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 6, leg_size = 6,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 6)) +
        scale_color_manual(values=host_order_color_dark) +
        guides(color="none")
  plot_scaff_detected_list[[magotu]] <- plot_scaff_detected
}
pdf("results/figures/10-instrain_plots/1-scaffolds_detected_by_magOTU.pdf", width = 20, height = 25)
marrangeGrob(plot_scaff_detected_list, nrow = 3, ncol = 3)
dev.off()
```

```{r instrain_plots_breadth_by_sample}
plot_breadth_by_sample_list <- list()
for (a_sample in unique(instrain_genomes_info_df$sample)) {
  plot_breadth_by_sample <- ggplot(instrain_genomes_info_df_all %>% mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100) %>%
                                filter(sample == a_sample),
                             aes(x = breadth,
                                 y = breadth_expected,
                                 color = Genus,
                              )
                            ) +
  geom_point() +
    ggtitle(paste0(a_sample)) +
      make_theme(theme_name=theme_few(), leg_pos="none",
                 # guide_nrow = 6,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_abline(slope=1, linetype = "dotted") +
        theme(strip.text = element_text(size = 6)) +
        scale_color_manual(values=genusColors)
  plot_breadth_by_sample_list[[a_sample]] <- plot_breadth_by_sample
}
pdf("results/figures/10-instrain_plots/1-breadth_vs_expected_by_sample.pdf", width = 20, height = 25)
marrangeGrob(plot_breadth_by_sample_list, nrow = 3, ncol = 3)
dev.off()
```


```{r instrain_plots_breadth_by_sample_with_coveraeg}
# redo using marrangeGrob !
plot_breadth_by_sample_w_cov_list <- list()
for (a_sample in unique(instrain_genomes_info_df$sample)) {
  plot_breadth_by_sample_w_cov <- ggplot(instrain_genomes_info_df_all %>% mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100) %>%
                                filter(sample == a_sample),
                             aes(y = breadth,
                                 x = breadth_expected,
                                 color = Genus,
                                 size = coverage
                              ), alpha = 0.75
                            ) +
  geom_point() +
    ggtitle(paste0(a_sample)) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 12,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_abline(slope=1, linetype = "dotted") +
        theme(strip.text = element_text(size = 6)) +
        scale_color_manual(values=genusColors) +
        guides(color = guide_legend(show = FALSE))
  plot_breadth_by_sample_w_cov_list[[a_sample]] <- plot_breadth_by_sample_w_cov
}
pdf("results/figures/10-instrain_plots/1-breadth_vs_expected_w_cov_by_sample.pdf", width = 20, height = 25)
marrangeGrob(plot_breadth_by_sample_w_cov_list, nrow = 3, ncol = 3)
dev.off()
```

```{r instrain_plots_breadth_by_magOTU}
plot_breadth_by_magOTU_list <- list()
for (magotu in unique(instrain_genomes_info_df_all)) {
  plot_breadth_by_magOTU <- ggplot(instrain_genomes_info_df_all %>% mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100),
                             aes(x = breadth,
                                 y = breadth_expected,
                                 color = Host,
                              )
                            ) +
                              geom_point() +
                                ggtitle(paste0(magotu)) +
                                  make_theme(theme_name=theme_few(), leg_pos="none",
                                            # guide_nrow = 6,
                                            setFill = F, setCol = F,
                                            x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                                            y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
                                    geom_abline(slope=1, linetype = "dotted") +
                                    theme(strip.text = element_text(size = 6)) +
                                    scale_color_manual(values=host_order_color_dark)
  plot_breadth_by_magOTU_list[[magotu]] <- plot_breadth_by_magOTU
}
pdf("results/figures/10-instrain_plots/1-breadth_vs_expected_by_magOTU.pdf", width = 20, height = 25)
marrangeGrob(plot_breadth_by_magOTU_list, nrow = 3, ncol = 3)
dev.off()
```

```{r instrain_plots_breadth_vs_num_reads}
# redo using marrangeGrob !
handmade_species_names <- read.csv("config/Species_MAG_Cluster-names.txt", header = T, sep = "\t") %>% 
                            mutate(cluster_name_instrain = paste0(MAG_species_Name_in_analysis, "-", RepMAG))
instrain_breadth_depth_unfiltered_all_handmade_handmainstrain_breadth_depth_unfiltered_all_handmade <- instrain_breadth_depth_unfiltered_all %>% 
                                          left_join(handmade_species_names %>% select(MAG_species_name_final, cluster_name_instrain), by = c("magOTU" = "cluster_name_instrain"))
instrain_breadth_depth_all_handmade <- instrain_breadth_depth_all %>% 
                                          left_join(handmade_species_names %>% select(MAG_species_name_final, cluster_name_instrain), by = c("magOTU" = "cluster_name_instrain"))


plot_breadth_num_reads <- ggplot() +
  geom_point(data = instrain_breadth_depth_unfiltered_all_handmade %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth
                              ), alpha = 0.4, color = "black"
                            ) +
  geom_point(data = instrain_breadth_depth_all_handmade %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                                 color = Host
                              )
                            ) +
  scale_y_continuous(trans = "log10", labels = comma) +
    facet_wrap_paginate(~MAG_species_name_final, ncol = 3, nrow = 3, page = 4) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 5,
                 setFill = F, setCol = F,
                 x_angle=90 ,x_vj=1, x_hj=1, x_size=0,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 10)) +
        geom_hline(yintercept = 0, linetype = "solid") +
        geom_hline(yintercept = 10, linetype = "dashed") +
        geom_hline(yintercept = 100, linetype = "dotted") +
        geom_vline(xintercept = 97, linetype = "dotted") +
        geom_vline(xintercept = 162, linetype = "dotted") +
        geom_vline(xintercept = 207, linetype = "dotted") +
        geom_vline(xintercept = 252, linetype = "dotted") +
        scale_size(range = c(1,4)) +
        scale_color_manual(values=host_order_color_dark)

paginate_save(plot_breadth_num_reads, "MAG_species_name_final", "results/figures/10-instrain_plots/2-mapped_reads_by_magOTU.pdf", pass_nrow = 3, pass_ncol = 3)
```
```{r instrain_plots_breadth_vs_coverage}
# redo using marrangeGrob


plot_breadth_num_reads <- ggplot() +
  geom_point(data = instrain_breadth_depth_unfiltered_all_handmade %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = coverage_median,
                                 size = breadth
                              ), alpha = 0.4, color = "black"
                            ) +
  geom_point(data = instrain_breadth_depth_all_handmade %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = coverage_median,
                                 size = breadth,
                                 color = Host
                              )
                            ) +
  scale_y_continuous(trans = "log10", labels = comma, limits = c(-0.1, NA)) +
    facet_wrap_paginate(~MAG_species_name_final, ncol = 3, nrow = 3, page = 4) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 5,
                 setFill = F, setCol = F,
                 x_angle=90 ,x_vj=1, x_hj=1, x_size=0,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 10)) +
        geom_hline(yintercept = 0, linetype = "solid") +
        geom_hline(yintercept = 10, linetype = "dashed") +
        geom_hline(yintercept = 100, linetype = "dotted") +
        geom_vline(xintercept = 97, linetype = "dotted") +
        geom_vline(xintercept = 162, linetype = "dotted") +
        geom_vline(xintercept = 207, linetype = "dotted") +
        geom_vline(xintercept = 252, linetype = "dotted") +
        scale_size(range = c(1,4)) +
        scale_color_manual(values=host_order_color_dark)

paginate_save(plot_breadth_num_reads, "MAG_species_name_final", "results/figures/10-instrain_plots/2-coverage_by_magOTU.pdf", pass_nrow = 3, pass_ncol = 3)
```

```{r instrain_plots_breadth_vs_num_reads_coverage}
# redo using marrangeGrob

plot_breadth_num_reads <- ggplot() +
  geom_point(data = instrain_breadth_depth_unfiltered_all_handmade %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                                 color = coverage_factor
                              )
                            ) +
  geom_point(data = instrain_breadth_depth_all_handmade %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                                 color = coverage_factor,
                              )
                            ) +
  geom_point(data = instrain_breadth_depth_all_handmade %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                              ), shape = 1, color = "black", stroke = 0.25
                            ) +
  scale_color_manual(values = c("<0.0001" = "#a50f15",
                                "0.0001 - 0.001" = "#d73027",
                                "0.001 - 0.01" = "#f46d43",
                                "0.01 - 0.1" = "#fdae61",
                                "0.1 - 1" = "#fee08b",
                                "1 - 10" = "#a6d96a", 
                                "10 - 100" = "#74c476", 
                                "100 - 1000" = "#41ab5d", 
                                ">1000" = "#238b45"
                              )
                          ) +
  scale_y_continuous(trans = "log10", labels = comma) +
    facet_wrap_paginate(~MAG_species_name_final, ncol = 3, nrow = 3, page = 4) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 9,
                 setFill = F, setCol = F, palettecolor = "RdYlGn",
                 x_angle=45 ,x_vj=1, x_hj=1, x_size=0,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 10)) +
        geom_hline(yintercept = 0, linetype = "solid") +
        geom_hline(yintercept = 10, linetype = "dashed") +
        geom_hline(yintercept = 100, linetype = "dotted") +
        geom_vline(xintercept = 97, linetype = "dotted") +
        geom_vline(xintercept = 162, linetype = "dotted") +
        geom_vline(xintercept = 207, linetype = "dotted") +
        geom_vline(xintercept = 252, linetype = "dotted") +
        scale_size(range = c(1,4))

paginate_save(plot_breadth_num_reads, "MAG_species_name_final", "results/figures/10-instrain_plots/2-mapped_reads_by_magOTU_w_coverage.pdf", pass_nrow = 3, pass_ncol = 3)
```

```{r coverage_nor_adj}
# plot_cov_df <- abundance_df %>%
#                 select(sample, Host, magOTU, coverage_median, coverage, coverage_adj, breadth) %>%
#                   pivot_longer(cols = c(coverage, coverage_adj, coverage_median), names_to = "cov_type", values_to = "coverage")
# plot_list <- list()
# for (magotu in unique(plot_cov_df$magOTU)) {
#   plot <- ggplot() +
#           geom_point(data = plot_cov_df %>% filter(magOTU == magotu),
#                              aes(x = factor(sample, samples),
#                                  y = coverage,
#                                 #  size = breadth,
#                                  shape = cov_type,
#                                  color = Host
#                               ), size = 3, alpha = 0.75
#                             ) +
#             scale_y_continuous(trans = "log10", labels = comma, limits = c(-0.1, NA)) +
#             make_theme(theme_name=theme_few(), leg_pos="right",
#                       guide_nrow = 5,
#                       setFill = F, setCol = F,
#                       x_angle=90 ,x_vj=1, x_hj=1, x_size=0,
#                       y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
#               scale_shape_manual(values = c("coverage" = 18, "coverage_adj" = 13, "coverage_median" = 15)) +
#               theme(strip.text = element_text(size = 10)) +
#               geom_hline(yintercept = 0, linetype = "solid") +
#               geom_hline(yintercept = 10, linetype = "dashed") +
#               geom_hline(yintercept = 100, linetype = "dotted") +
#               geom_vline(xintercept = 97, linetype = "dotted") +
#               geom_vline(xintercept = 162, linetype = "dotted") +
#               geom_vline(xintercept = 207, linetype = "dotted") +
#               geom_vline(xintercept = 252, linetype = "dotted") +
#               scale_size(range = c(1,4)) +
#               # scale_color_manual(values=c("#bebada","#fb8072", "#80b1d3"))
#               scale_color_manual(values=host_order_color_dark)
#   plot_list[[magotu]] <- plot
# }

# pdf("results/figures/10-instrain_plots/2-coverage_by_magOTU_w_coverage.pdf", width = 12, height = 15)
# marrangeGrob(plot_list, nrow = 5, ncol = 1)
# dev.off()
```


# Reads mapping to database

```{r}
colnames(df_reads_updated)

ggplot(df_reads_plot %>% 
          filter(!is.na(Number)) %>% 
          filter(Sample %in% samples_MY) %>% 
          filter(Type %in% c("Total_clean", "MAGs_DB_mf", "Host_mapped_hf")),
          aes(y = factor(Sample, samples_MY), x = Number, fill = Type)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Host, scale = "free_y") +
  scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    make_theme(leg_pos = "right", guide_nrow = 3)
ggsave("results/figures/10-instrain_plots/3-reads_mapped_to_each_db.pdf", width = 12, height = 15)

# of the non-host reads how much maps to mags db
df_reads_mod <- df_reads %>%
                  filter(Sample %in% samples_IN_MY) %>%
                  mutate(nonhost_hf = Total_clean - Host_mapped_hf) %>%
                  mutate(perc_nonhost_mag_hf = MAGs_DB_hf/nonhost_hf*100) %>%
                  mutate(perc_unmapped_nonhost_hf = 100 - perc_nonhost_mag_hf) %>%
                  mutate(MAG_DB_mapped = perc_nonhost_mag_hf) %>%
                  mutate(unmapped = perc_unmapped_nonhost_hf) %>%
                  select(MAG_DB_mapped, unmapped, Sample, nonhost_hf) %>%
                  pivot_longer(cols = c(MAG_DB_mapped, unmapped, nonhost_hf), names_to = "Type", values_to = "Number") %>%
                  mutate(Host = Vectorize(get_host_from_sample_name)(Sample))

ggplot() +
  geom_bar(data = df_reads_mod %>% filter(Type != "nonhost_hf"),
            aes(x = Number, y = factor(Sample, samples_IN_MY), fill = factor(Type, c("unmapped", "MAG_DB_mapped"))),
            stat = "identity", position = "stack") +
  facet_wrap(~Host, scale = "free_y") +
  labs(fill = "Non-host reads", x = "Percentage", y = "Sample") +
  scale_fill_manual(values = c("MAG_DB_mapped" = "#ccebc5", "unmapped" = "#fbb4ae"), labels = c("MAG_DB_mapped" = "Mapped to MAGs", "unmapped" = "Unmapped to all")) +
  scale_x_continuous(labels=unit_format(unit = "%", scale = 1)) +
  new_scale_fill() +
  geom_tile(data = df_reads_mod %>% filter(Type == "nonhost_hf"), 
            aes(x = -8, width = 10, y = factor(Sample, samples_IN_MY), fill = Number), color = "#000000") +
    scale_fill_gradient2(low = "#f7fbff", mid = "#4292c6", high = "#08306b", midpoint = 4e+07,
                         label = scientific_10) +
    labs(fill = "Number of non-host reads") +
    geom_vline(xintercept = 80, linetype = "dotted") +
    make_theme(setCol = F, setFill = F, leg_pos = "right",
                guide_nrow = 3, modify_guide = F,
                x_angle = 45, x_vj = 1, x_hj = 1)
ggsave("results/figures/10-instrain_plots/3-perc_mapped_to_mag_vs_unmapped.pdf", width = 12, height = 15)

# plot qpcr Ct values against number of mag_db_mf and mag_db_hf

df_plot_ct <- df_reads_plot %>% 
                filter(!is.na(Number)) %>% 
                filter(Sample %in% samples_MY) %>% 
                filter(Type %in% c("Total_clean", "MAGs_DB_mf", "Host_mapped_hf")) %>%
                pivot_wider(names_from = Type, values_from = Number) %>%
                left_join(qpcr_plot_df %>% select(Ct_mean, Sample.Name), by = c("Sample" = "Sample.Name"))

ggplot(df_plot_ct) +
  geom_point(aes(y = MAGs_DB_mf, x = Ct_mean, color = Host), size = 3) +
  labs(x = "Ct value (universal 16S)", y = "Number of reads mapped to MAGs") +
  scale_x_reverse() +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_color_manual(values=host_order_color_dark) +
    make_theme(leg_pos = "right", guide_nrow = 5, setFill = F, setCol = F)
ggsave("results/figures/10-instrain_plots/3-reads_mapped_to_mag_vs_ct.pdf", width = 12, height = 15)

ggplot(df_plot_ct) +
  geom_point(aes(y = Host_mapped_hf, x = Ct_mean, color = Host), size = 3) +
  scale_x_reverse() +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_color_manual(values=host_order_color_dark) +
    make_theme(leg_pos = "right", guide_nrow = 5, setFill = F, setCol = F)
ggsave("results/figures/10-instrain_plots/3-reads_mapped_to_host_vs_ct.pdf", width = 12, height = 15)

ggplot(df_plot_ct) +
  geom_point(aes(y = Total_clean, x = Ct_mean, color = Host), size = 3) +
  scale_x_reverse() +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_color_manual(values=host_order_color_dark) +
    make_theme(leg_pos = "right", guide_nrow = 5, setFill = F, setCol = F)
ggsave("results/figures/10-instrain_plots/3-reads_sequenced_vs_ct.pdf", width = 12, height = 15)

```

# Estimate host range
```{r}
host_comp_order <- c("mellifera_mellifera",
                    "cerana_cerana",
                    "dorsata_dorsata",
                    "florea_florea",
                    "andreniformis_andreniformis",
                    "andreniformis_florea",
                    "cerana_mellifera",
                    "dorsata_mellifera",
                    "dorsata_florea",
                    "cerana_dorsata",
                    "andreniformis_dorsata",
                    "andreniformis_cerana",
                    "cerana_florea",
                    "florea_mellifera",
                    "andreniformis_mellifera")

# df_shared <- read.csv("results/figures/magotu_shared_same_diff.csv", stringsAsFactors = F) %>%
#               pivot_longer(cols = !c("X"), names_to = "host_comb", values_to = "prop_shared") %>%
#               left_join(magOTU_tax_info, by = c("X" = "magOTU"))

# p_list <- list()
# for (genus in unique(df_shared$Genus)) {
#   p <- ggplot(df_shared %>% filter(Genus == genus),
#   aes(x = factor(host_comb, host_comp_order), y = prop_shared, group = X)
#   ) +
#   geom_point(color = "#000000", size = 3) +
#   geom_point(aes(color = Genus), size = 2) +
#   scale_color_manual(values=genusColors) +
#   new_scale_color() +
#   geom_line(aes(color = X), size = 1) +
#   scale_color_manual(values=head(colorRampPalette(c(genusColors[[genus]], "#FFFFFF"))(2+length(unique(df_shared %>% filter(Genus == genus) %>% pull(X)))), -2)) +
#     labs(x = "Host species pair", y = "") +
#     ggtitle(paste0(genus, " - Proportion of host species pairs that share the magOTU")) +
#     geom_vline(xintercept = 5.5, linetype = "solid", color = "#000000") +
#     make_theme(setFill = F, setCol = F, leg_pos = "none",
#                y_hj = 1, y_vj = 1, y_size = 7,
#                x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
#     ) +
#     ylim(0,1) +
#       theme(plot.title = element_text(size = 10, face = "bold"),
#             panel.background = element_rect(fill = "white", colour = "black"),
#             panel.grid.major = element_line(colour = "grey", linetype = "solid"),
#       )  
#   p_list[[genus]] <- p
# }
# pdf("results/figures/3-magotu_shared.pdf", width = 12, height = 15)
# marrangeGrob(p_list, nrow = 5, ncol = 2)
# dev.off()


# df_shared_ab <- read.csv("results/figures/magotu_shared_same_diff_ab.csv", stringsAsFactors = F) %>%
#               pivot_longer(cols = !c("X"), names_to = "host_comb", values_to = "prop_shared") %>%
#               left_join(magOTU_tax_info, by = c("X" = "magOTU"))

# p_list <- list()
# for (genus in unique(df_shared_ab$Genus)) {
#   p <- ggplot(df_shared_ab %>% filter(Genus == genus),
#   aes(x = factor(host_comb, host_comp_order), y = prop_shared, group = X)
#   ) +
#   geom_point(color = "#000000", size = 3) +
#   geom_point(aes(color = Genus), size = 2) +
#   scale_color_manual(values=genusColors) +
#   new_scale_color() +
#   geom_line(aes(color = X), size = 1) +
#   scale_color_manual(values=head(colorRampPalette(c(genusColors[[genus]], "#FFFFFF"))(2+length(unique(df_shared_ab %>% filter(Genus == genus) %>% pull(X)))), -2)) +
#     labs(x = "Host species pair", y = "") +
#     ggtitle(paste0(genus, " - Proportion of host species pairs that share the magOTU")) +
#     geom_vline(xintercept = 5.5, linetype = "solid", color = "#000000") +
#     make_theme(setFill = F, setCol = F, leg_pos = "none",
#                y_hj = 1, y_vj = 1, y_size = 7,
#                x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
#     ) +
#     ylim(0,1) +
#       theme(plot.title = element_text(size = 10, face = "bold"),
#             panel.background = element_rect(fill = "white", colour = "black"),
#             panel.grid.major = element_line(colour = "grey", linetype = "solid"),
#       )  
#   p_list[[genus]] <- p
# }
# pdf("results/figures/3-magotu_shared_ab.pdf", width = 12, height = 15)
# marrangeGrob(p_list, nrow = 5, ncol = 2)
# dev.off()



# rhodes specificity index (with modification for std.)
# https://hal.sorbonne-universite.fr/hal-03246067/document
# implemented this in python and made "results/figures/magotu_rohdes_index.csv"

prevs_df <- read.csv("results/figures/magotu_prevs.csv") %>%
              pivot_longer(cols = !c("X"), names_to = "host", values_to = "prevalence_in_host") %>%
              mutate(Host = paste0("Apis ", host))

handmade_species_names <- read.csv("config/Species_MAG_Cluster-names.txt", header = T, sep = "\t") %>% 
                            mutate(cluster_name_instrain = paste0(MAG_species_Name_in_analysis, "-", RepMAG))

df_plot <- abundance_df %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            mutate(detected = ifelse(coverage > 1 & breadth > breadth_cutoff, 1, 0)) %>%
                            select(sample, Host, magOTU, detected, Genus, coverage, copies_cov, rel_cov) %>%
                            # filter(detected == 1) %>% # only using the breadth cut-off ... adding cov cut-off did not seem to change much
                            left_join(prevs_df, by = c("magOTU" = "X", "Host" = "Host")) %>%
                            select(magOTU, coverage, copies_cov, rel_cov, prevalence_in_host, Genus, Host, sample) %>%
                            left_join(handmade_species_names %>% select(MAG_species_name_final, cluster_name_instrain), by = c("magOTU" = "cluster_name_instrain"))

df_rohdes <- read.csv("results/figures/magotu_rohdes_index.csv", stringsAsFactors = F) %>%
              left_join(magOTU_tax_info, by = c("X" = "magOTU")) %>%
              pivot_longer(cols = !c("X", "Genus"), names_to = "type", values_to = "rohdes_index") %>%
              mutate(specific = ifelse(rohdes_index == 1, "specific", "general")) %>%
              left_join(df_plot %>% select(MAG_species_name_final, magOTU) %>% unique, by = c("X" = "magOTU")) %>%
              mutate(X_old = X) %>%
              mutate(X = MAG_species_name_final)
write.csv(df_rohdes, "results/figures/magotu_rohdes_index_annotated.csv", row.names = F, quote = F)
p_list <- list()
for (genus in unique(df_rohdes$Genus)) {
  p <- ggplot(df_rohdes %>% filter(Genus == genus)
  ) +
  geom_bar(aes(x = X, 
               y = rohdes_index,
               fill = type, 
               color = factor(specific, c("specific", "general"))), size = 0.25, stat = "identity", position = position_dodge2(padding = 0.2)) +
    labs(x = "magOTU", y = "Rohde's index of host specificity", fill = "Input Type", color = "specificity") +
    ggtitle(paste0(genus, " - Rhodes specificity index")) +
    scale_color_manual(values=c("general" = "#ffffff","specific" = "#000000")) +
    make_theme(setFill = T, setCol = F, leg_pos = "right",
               y_hj = 1, y_vj = 1, y_size = 7, leg_size = 6,
               x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
    ) +
      theme(plot.title = element_text(size = 10, face = "bold"),
            panel.background = element_rect(fill = "white", colour = "black"),
            panel.grid.major = element_line(colour = "grey", linetype = "solid"),
      )
  p_list[[genus]] <- p
}
pdf("results/figures/3-magotu_rohdes_types.pdf", width = 12, height = 15)
marrangeGrob(p_list, nrow = 5, ncol = 2)
dev.off()

# Use prevalence, it seems to be the most robust... compare more quantitatively ???...

magotu_order <- df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% arrange(Genus, rohdes_index) %>% pull(X)
ggplot(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__")
  ) +
  geom_bar(aes(y = factor(X, magotu_order), 
               x = rohdes_index,
              #  color = factor(specific, c("specific", "general")),
               fill = Genus
              ), 
              color = "#000000",
              size = 0.25, stat = "identity", position = position_dodge2(padding = 0.2)) +
    labs(y = "magOTU", x = "Rohde's index of host specificity", fill = "Input Type", color = "specificity") +
    scale_color_manual(values=c("general" = "#ffffff","specific" = "#000000")) +
    scale_fill_manual(values=genusColors) +
    facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
    make_theme(setFill = F, setCol = F, leg_pos = "none",
               y_hj = 1, y_vj = 1, y_size = 7, leg_size = 6, guide_nrow = 22,
               x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
    ) +
      theme(plot.title = element_text(size = 10, face = "bold"),
            strip.text.y = element_text(angle = 0),
            panel.background = element_rect(fill = "white", colour = "black"),
            panel.grid.major = element_line(colour = "grey", linetype = "solid")
      )
 ggsave("results/figures/3-magotu_rohdes.pdf", width = 12, height = 15)


magotu_order <- df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% arrange(Genus, rohdes_index) %>% pull(X)
p_rohdes <- ggplot(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__")) +
            geom_bar(aes(y = factor(X, magotu_order), 
                         x = rohdes_index
                         ),
                     fill = "#f6e8c3",
                     color = "#000000",
                     size = 0.25, stat = "identity", position = position_dodge2(padding = 0.2)) +
    labs(y = "magOTU", x = "Rohde's index of host specificity", color = "specificity") +
    # scale_fill_manual(values=genusColors) +
    facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
    make_theme(setFill = F, setCol = F, leg_pos = "none",
               y_hj = 1, y_vj = 1, y_size = 7, leg_size = 6, guide_nrow = 22,
               x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
    ) +
      theme(plot.title = element_text(size = 10, face = "bold"),
            strip.text.y = element_text(angle = 0),
            panel.background = element_rect(fill = "white", colour = "black"),
            panel.grid.major = element_line(colour = "grey", linetype = "solid")
      )
p_hp_m <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>% 
                                      unique %>% filter(Host == "Apis mellifera"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis mellifera")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis mellifera"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))
p_hp_c <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>%
                                      unique %>% filter(Host == "Apis cerana"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis cerana")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis cerana"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))
p_hp_d <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>% 
                                      unique %>% filter(Host == "Apis dorsata"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis dorsata")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis dorsata"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))
p_hp_f <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>% 
                                      unique %>% filter(Host == "Apis florea"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis florea")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis florea"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))
p_hp_a <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>%
                                      unique %>% filter(Host == "Apis andreniformis"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis andreniformis")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis andreniformis"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))

pdf("results/figures/3-magotu_rohdes_with_host.pdf", width = 12, height = 15)
grid.arrange(p_rohdes, p_hp_m, p_hp_c, p_hp_d, p_hp_f, p_hp_a, nrow = 1, widths = c(7,1,1,1,1,1))
dev.off()

df_rohdes_by_host <- df_rohdes %>%
                      filter(type == "prev") %>%
                      select(X_old, X, rohdes_index) %>%
                      left_join(
                        read.csv("results/figures/magotu_ab_ranks.csv") %>%
                          pivot_longer(!X, names_to = "Host", values_to = "rank"),
                        by = c("X_old" = "X")
                      ) %>%
                      mutate(Host = paste0("Apis ", Host)) %>%
                      mutate(rank = as.factor(rank)) %>%
                      # filter(rank == "1") %>%
                      filter(X_old %in% (prevs_df %>% filter(prevalence_in_host >= 0.5) %>% pull(X))) %>%
                      left_join(df_plot %>% select(Genus, MAG_species_name_final) %>% unique, by = c("X" = "MAG_species_name_final"))


ggplot() +
  # geom_violin(data = df_rohdes_by_host,
  geom_boxplot(data = df_rohdes_by_host %>% filter(rank %in% c("1")),
       aes(x = factor(Host, levels = c("Apis mellifera", "Apis dorsata", "Apis cerana", "Apis florea", "Apis andreniformis")),
           y = rohdes_index,
          #  fill = factor(Host, levels = c(host_order)),
         ), alpha = 0.6) +
  # geom_jitter(data = df_rohdes_by_host,
  geom_jitter(data = df_rohdes_by_host %>% filter(rank %in% c("1")),
       aes(x = factor(Host, levels = c("Apis mellifera", "Apis dorsata", "Apis cerana", "Apis florea", "Apis andreniformis")),
           y = rohdes_index,
          #  fill = factor(Host, levels = c(host_order)),
          fill = Genus
         ), width=0.1, size = 4, shape = 21, color = "#000000") +
  labs(y = "Rohde's index", x = "Host species (maximum prevalence)", fill = "Host") +
  scale_color_manual(values=genusColors) +
  scale_fill_manual(values=genusColors) +
  scale_size_manual(values=c("1" = 4,
                              "2" = 2,
                              "3" = 2,
                              "4" = 2,
                              "5" = 2
  )) +
  scale_alpha_manual(values=c("1" = 1,
                              "2" = 0.5,
                              "3" = 0.5,
                              "4" = 0.5,
                              "5" = 0.5
  )) +
    make_theme(leg_pos = "right", setFill = F, setCol = F, x_angle = 30, x_vj = 1, x_hj = 1, 
               x_size = 15, y_size = 15, axis_x_title = 20, axis_y_title = 20,
               guide_nrow = 20
              )
      ggsave("results/figures/3-magotu_rohdes_by_magotu_across.pdf", width = 12, height = 15)


df_rohdes_across_hosts_plot <- abundance_df %>% filter(!is.na(magOTU)) %>%
                            filter(!sample %in% samples_to_exclude) %>%
                            filter(breadth > breadth_cutoff) %>%
                            mutate(detected = ifelse(coverage > 1 & breadth > breadth_cutoff, 1, 0)) %>%
                            select(magOTU, Genus, Host, detected) %>%
                            filter(detected == 1) %>%
                            left_join(df_rohdes %>% 
                                        filter(type == "prev") %>%
                                        select(X_old, rohdes_index, MAG_species_name_final), 
                                        by = c("magOTU" = "X_old")) %>%
                            filter(!is.na(rohdes_index)) %>%
                            mutate(Species = MAG_species_name_final) %>%
                            select(Species, rohdes_index, Host, detected) %>%
                            group_by(Species, Host) %>%
                            summarize(rohdes_index, total = sum(detected)) %>%
                            unique() %>%
                            filter(total > 3)

ggplot() +
  # geom_violin(data = df_rohdes_by_host %>% filter(rank %in% c("1")),
  # geom_boxplot(data = df_rohdes_by_host %>% filter(rank %in% c("1", "2")),
  geom_boxplot(data = df_rohdes_across_hosts_plot,
       aes(x = factor(Host, levels = c("Apis mellifera", "Apis dorsata", "Apis cerana", "Apis florea", "Apis andreniformis")),
           y = rohdes_index,
           fill = factor(Host, levels = c(host_order)),
         ), alpha = 0.5, outlier.shape = NA) +
  # geom_jitter(data = df_rohdes_by_host,
  geom_jitter(data = df_rohdes_across_hosts_plot,
  # geom_jitter(data = df_rohdes_by_host %>% filter(rank %in% c("1")),
       aes(x = factor(Host, levels = c("Apis mellifera", "Apis dorsata", "Apis cerana", "Apis florea", "Apis andreniformis")),
           y = rohdes_index#,
          #  fill = factor(Host, levels = c(host_order)),
          # size = total
         ), width=0.1, size = 3, color = "#000000", alpha = 0.75) +
  labs(y = "Rohde's index", x = "Host species (maximum prevalence)", fill = "Host") +
  scale_fill_manual(values=host_order_color) +
    make_theme(leg_pos = "right", setFill = F, setCol = F, x_angle = 30, x_vj = 1, x_hj = 1, 
               x_size = 15, y_size = 15, axis_x_title = 20, axis_y_title = 20,
               guide_nrow = 20
              )
      ggsave("results/figures/3-magotu_rohdes_by_magotu_across.pdf", width = 12, height = 10)


ggplot() +
  # geom_violin(data = df_rohdes_by_host %>% filter(rank %in% c("1")),
  geom_boxplot(data = df_rohdes_by_host %>% filter(rank %in% c("1")),
       aes(x = factor(Genus, genera),
           y = rohdes_index,
          #  fill = factor(Host, levels = c(host_order)),
         ), alpha = 0.6) +
  # geom_jitter(data = df_rohdes_by_host,
  geom_jitter(data = df_rohdes_by_host %>% filter(rank %in% c("1")),
       aes(fill = factor(Host, levels = c("Apis mellifera", "Apis dorsata", "Apis cerana", "Apis florea", "Apis andreniformis")),
           y = rohdes_index,
          #  fill = factor(Host, levels = c(host_order)),
          x = Genus
         ), width=0.1, size = 6, shape = 21, color = "#000000", alpha = 0.6) +
  labs(y = "Rohde's index", x = "Genus", fill = "Genus", color = "Host") +
  scale_color_manual(values=host_order_color_dark) +
  scale_fill_manual(values=host_order_color_dark) +
  scale_size_manual(values=c("1" = 4,
                              "2" = 2,
                              "3" = 2,
                              "4" = 2,
                              "5" = 2
  )) +
  scale_alpha_manual(values=c("1" = 1,
                              "2" = 0.5,
                              "3" = 0.5,
                              "4" = 0.5,
                              "5" = 0.5
  )) +
    make_theme(leg_pos = "right", setFill = F, setCol = F, x_angle = 30, x_vj = 1, x_hj = 1, 
               x_size = 15, y_size = 15, axis_x_title = 20, axis_y_title = 20,
               guide_nrow = 20
              )
      ggsave("results/figures/3-magotu_rohdes_by_magotu_across_genera.pdf", width = 20, height = 15)

# set.seed(123)
# long <- rnorm(50, sd=100)
# lat <- rnorm(50, sd=50)
# d <- data.frame(long=long, lat=lat)
# d <- with(d, d[abs(long) < 150 & abs(lat) < 70,])
# n <- nrow(d)
# d$region <- factor(1:n)
# d$A <- abs(rnorm(n, sd=1))
# d$B <- abs(rnorm(n, sd=2))
# d$C <- abs(rnorm(n, sd=3))
# d$D <- abs(rnorm(n, sd=4))
# d[1, 4:7] <- d[1, 4:7] * 3
# head(d)

# ggplot() + geom_scatterpie(aes(x=long, y=lat, group=region), data=d,
#                            cols=LETTERS[1:4]) + coord_equal()


# using scatter pie plot
# make one column per host and the number of bees in that host that have the bacteria / total number of bees (should add up to one) is the values
# then plot the values as a pie chart in the scatter plot
# the size of the pie chart should be the number of bees that have the bacteria
# 

df_scatter_pie <- abundance_df %>% filter(!is.na(magOTU)) %>%
                            filter(!sample %in% samples_to_exclude) %>%
                            filter(breadth > breadth_cutoff) %>%
                            mutate(detected = ifelse(coverage > 1 & breadth > breadth_cutoff, 1, 0)) %>%
                            select(magOTU, Genus, Host, detected) %>%
                            filter(detected == 1) %>%
                            pivot_wider(c(magOTU, Genus), names_from = Host, values_from = detected, values_fn = sum) %>%
                            left_join(df_rohdes %>% 
                                        filter(type == "prev") %>%
                                        select(X_old, rohdes_index, MAG_species_name_final), 
                                        by = c("magOTU" = "X_old")) %>%
                            mutate(rohdes_index = as.numeric(rohdes_index)) %>%
                            mutate(rohdes_index = 100*rohdes_index) %>%
                            select(-magOTU) %>%
                            # mutate(mellifera = `Apis mellifera`) %>%
                            # select(-`Apis mellifera`) %>%
                            # mutate(cerana = `Apis cerana`) %>%
                            # select(-`Apis cerana`) %>%
                            # mutate(dorsata = `Apis dorsata`) %>%
                            # select(-`Apis dorsata`) %>%
                            # mutate(florea = `Apis florea`) %>%
                            # select(-`Apis florea`) %>%
                            # mutate(andreniformis = `Apis andreniformis`) %>%
                            # select(-`Apis andreniformis`) %>%
                            mutate(across(-c(MAG_species_name_final, Genus), ~replace_na(., 0))) %>%
                            group_by(MAG_species_name_final) %>%
                            mutate(total = `Apis mellifera` +
                                            `Apis cerana` +
                                            `Apis dorsata` +
                                            `Apis florea` +
                                            `Apis andreniformis`
                            ) %>%
                            filter(total > 5) %>%
                            select(-total) %>%
                            filter(!is.na(MAG_species_name_final)) %>%
                            filter(!is.na(Genus)) %>%
                            ungroup() %>%
                            mutate(Species = as.factor(MAG_species_name_final)) %>%
                            select(-MAG_species_name_final)
# spec_list <- df_scatter_pie %>% pull(Species) %>% unique
# i = 0
# spec_names_num <- 
# for (spec in spec_list) {

# }
# df_scatter_pie %>% pull(Genus) %>% unique
genus_names_num <-  c(
  "g__Bombilactobacillus" = 1,
  "g__Lactobacillus" = 2,
  "g__CALYQJ01" = 3,
  "g__Bifidobacterium" = 4,
  "g__Gilliamella" = 5,
  "g__CALYQQ01" = 6,
  "g__Snodgrassella" = 7,
  "g__Bartonella_A" = 8,
  "g__Frischella" = 9,
  "g__Apibacter" = 10,
  "g__Commensalibacter" = 11,
  "g__Dysgonomonas" = 12,
  "g__Pectinatus" = 13,
  "g__Entomomonas" = 14,
  "g__WRHT01" = 15,
  "g__Saezia" = 16,
  "g__Apilactobacillus" = 17,
  "g__Enterobacter" = 18,
  "g__Parolsenella" = 19,
  "g__Pantoea" = 20,
  "g__Melissococcus" = 21,
  "g__Spiroplasma" = 22,
  "g__Klebsiella" = 23,
  "g__Escherichia" = 24,
  "g__Bombella" = 25,
  "g__Q3-R57-64" = 26,
  "g__Corynebacterium" = 27,
  "g__Pluralibacter" = 28
)
get_genus_num <- function(genus) {
  if (is.na(genus)) {
    return(0)
  }
  genus = as.character(genus)
  # print(genus)
  return(genus_names_num[[genus]])
}
get_genus_name <- function(num, mult = 5) {
  num = num / mult
  for (name in names(genus_names_num)) {
    if (genus_names_num[[name]] == num) {
      return(name)
    }
  }
}
log(134)
log(8)


df_scatter_pie_plot <- df_scatter_pie %>%
                        # mutate(Species = as.numeric(Species)) %>%
                            mutate(Genus = Vectorize(get_genus_num)(Genus)) %>%
                            # mutate(Genus = as.numeric(Genus)) %>%
                            mutate(Genus = Genus * 15) %>%
                            group_by(Species) %>%
                            mutate(total = `Apis mellifera` +
                                            `Apis cerana` +
                                            `Apis dorsata` +
                                            `Apis florea` +
                                            `Apis andreniformis`
                            ) %>%
                            # mutate(`Apis mellifera` = `Apis mellifera` / total,
                            #        `Apis cerana` = `Apis cerana` / total,
                            #        `Apis dorsata` = `Apis dorsata` / total,
                            #        `Apis florea` = `Apis florea` / total,
                            #        `Apis andreniformis` = `Apis andreniformis` / total
                            # ) %>%
                            # log transform it and make it be in the range of 0.01
                            mutate(total = log(total)) %>% View
                            # # randomly add or subtract a number between 0 and 0.2 from each entry in Genus column
                            mutate(Genus = Genus + runif(length(Genus), -0.1,0.1))

list_nums <- seq(15, length(genus_names_num)*15, 15)
list_names <- lapply(list_nums, function(x) get_genus_name(x,15))
names(list_names) <- list_nums

ggplot() +
  geom_scatterpie(aes(x = Genus,
                      y = rohdes_index,
                      r = total,
                      group = Species),
                  d = df_scatter_pie_plot,
                  cols = host_order,
                  size=0.1, 
                  alpha = 0.75) +
  scale_fill_manual(values = host_order_color_dark) +
  make_theme(leg_pos = "bottom", setFill = F, setCol = F,
               y_size = 15, x_size = 8, axis_x_title = 20, axis_y_title = 20,
               x_angle = 90, modify_guide = F,
               x_vj = 1, x_hj = 1,
              ) +
  scale_x_continuous(breaks =  list_nums, labels = list_names) +
  geom_scatterpie_legend(c(log(5), log(50), log(150)),
                          x = 20, y = 0,
                         labeller = function(x) round(exp(x))
  ) +
  coord_equal()
  ggsave("results/figures/3-magotu_rohdes_scatter_pie_radius.pdf", width = 15, height = 10)

ggplot() +
  geom_scatterpie(aes(x = Genus,
                      y = rohdes_index,
                      # r = total,
                      group = Species),
                  d = df_scatter_pie_plot,
                  cols = host_order,
                  size=0.1, 
                  alpha = 0.75) +
  scale_fill_manual(values = host_order_color_dark) +
  make_theme(leg_pos = "bottom", setFill = F, setCol = F,
               y_size = 15, x_size = 8, axis_x_title = 20, axis_y_title = 20,
               x_angle = 90,
               x_vj = 1, x_hj = 1,
              ) +
  scale_x_continuous(breaks =  list_nums, labels = list_names) +
  coord_equal()
  ggsave("results/figures/3-magotu_rohdes_scatter_pie.pdf", width = 15, height = 10)

# resume HERE

df_multi_pie <- df_scatter_pie %>%
                  pivot_longer(cols = !c("rohdes_index", "Genus", "Species"), names_to = "Host", values_to = "detected") %>%
                  filter(detected > 0) %>%
                  group_by(Species, Genus) %>%
                  mutate(total = sum(detected))

ggplot() +
  geom_point(aes(y = Species,
                 x = rohdes_index,
                #  group = Species,
                 color = Host,
                 size = detected), 
            data = df_multi_pie,
            alpha = 0.75) +
  # coord_polar("y", start = 0) +
  # facet_wrap(~Genus) +
  # scale_fill_manual(values = host_order_color_dark) +
  scale_color_manual(values = host_order_color_dark)
  ggsave("results/figures/3-magotu_rohdes_species_scatter_size.pdf", width = 15, height = 10)

ggplot() +
  geom_point(aes(y = Genus,
                 x = rohdes_index,
                #  group = Species,
                 color = Host,
                 size = detected), 
            data = df_multi_pie,
            alpha = 0.75) +
  # coord_polar("y", start = 0) +
  # facet_wrap(~Genus) +
  scale_fill_manual(values = host_order_color_dark) +
  scale_color_manual(values = host_order_color_dark)
  ggsave("results/figures/3-magotu_rohdes_genus_scatter_size.pdf", width = 15, height = 10)



abundance_df %>% select(sample, coverage) %>% filter(coverage > 0) %>% pull(sample) %>% unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis mellifera") %>%
  pull(sample) %>% 
  unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis cerana") %>%
  pull(sample) %>% 
  unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis dorsata") %>%
  pull(sample) %>% 
  unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis florea") %>%
  pull(sample) %>% 
  unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis andreniformis") %>%
  pull(sample) %>% 
  unique %>% length

# abundance_df %>% select(sample, coverage, Host, magOTU) %>%
#   filter(coverage > 0) %>%
#   filter(Host == "Apis florea") %>% 
#   filter(grepl("g__Dysgono", magOTU)) %>% 
#   view

# data.frame(cbind(magOTU = magotu_order)) %>% 
#                             left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X), by = c(magOTU = "X")) %>%
#                             left_join(df_plot %>% 
#                             select(prevalence_in_host, Host, magOTU)) %>%
#                             view


# plot rohdes (prev) vs. abundance per magOTU 
# (each magOTU is at the same prevalence regardless of what sample it comes from)
# the ones at high prevalence will also have more points

p_list_cov_prev <- list()
for (magotu in unique(df_plot$magOTU)) {
  p <- ggplot(df_plot %>% left_join(df_rohdes %>% 
  select(!Genus) %>%
  filter(type == "prev"), by = c("magOTU" = "X")) %>% filter(magOTU == magotu),
  aes(y = prevalence_in_host, x = copies_cov, color = Host)
  ) +
  geom_point(size = 3) +
  labs(y = "Prevalence in host", x = "Coverage (qPCR normalized)") +
  scale_color_manual(values=host_order_color_dark) +
  make_theme(setFill = F, setCol = F, leg_pos = "none",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             axis_y_title = 10
  ) +
  scale_x_continuous(trans = "log10") +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
  p_list_cov_prev[[magotu]] <- p
}

pdf("results/figures/3-magotu_prev_vs_abundance_copies.pdf", width = 12, height = 15)
marrangeGrob(p_list_cov_prev, nrow = 5, ncol = 2)
dev.off()

p_list_cov_prev <- list()
for (magotu in unique(df_plot$magOTU)) {
  p <- ggplot(df_plot %>% left_join(df_rohdes %>% 
  select(!Genus) %>%
  filter(type == "prev"), by = c("magOTU" = "X")) %>% filter(magOTU == magotu),
  aes(y = prevalence_in_host, x = rel_cov, color = Host)
  ) +
  geom_point(size = 3) +
  geom_vline(xintercept = 0.01, linetype = "dashed") +
  geom_vline(xintercept = 0.1, linetype = "dashed") +
  labs(y = "Prevalence in host", x = "Relative coverage") +
  scale_color_manual(values=host_order_color_dark) +
  make_theme(setFill = F, setCol = F, leg_pos = "none",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             axis_y_title = 10
  ) +
  scale_x_continuous(trans = "log10") +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
  p_list_cov_prev[[magotu]] <- p
}

pdf("results/figures/3-magotu_prev_vs_abundance_rel_cov.pdf", width = 12, height = 15)
marrangeGrob(p_list_cov_prev, nrow = 5, ncol = 2)
dev.off()


ggplot(df_plot %>% left_join(df_rohdes %>% 
  select(!Genus) %>%
  filter(type == "prev"), by = c("magOTU" = "X")),
  aes(y = prevalence_in_host, x = rel_cov, color = Genus, group = sample)
  ) +
  facet_wrap(~Host, ncol = 2) +
  geom_point(size = 3) +
  geom_vline(xintercept = 0.01, linetype = "dashed") +
  geom_vline(xintercept = 0.1, linetype = "dashed") +
  labs(y = "Prevalence in host", x = "Relative coverage") +
  scale_color_manual(values=genusColors) +
  make_theme(setFill = F, setCol = F, leg_pos = "none",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             axis_y_title = 10
  ) +
  scale_x_continuous(trans = "log10") +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
ggsave("results/figures/3-magotu_prev_vs_rel_cov.pdf", width = 12, height = 15)

# Could you look at the abundance of the highly specific ones vs the non-specific ones. Do the non-specific ones tend to have a preferred host, in which they are most abundant?
ggplot(df_plot %>% left_join(df_rohdes %>% 
  select(!Genus) %>%
  filter(type == "prev"), by = c("magOTU" = "X_old")),
  aes(y = rohdes_index, x = rel_cov, color = Genus, group = sample)
  ) +
  geom_point(size = 3) +
  geom_vline(xintercept = 0.01, linetype = "dashed") +
  geom_vline(xintercept = 0.1, linetype = "dashed") +
  geom_vline(xintercept = 0, linetype = "solid") +
  facet_wrap(~Host, ncol = 2) +
  labs(y = "Rohdes index", x = "Relative coverage") +
  scale_color_manual(values=genusColors) +
  make_theme(setFill = F, setCol = F, leg_pos = "none",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             axis_y_title = 10
  ) +
  scale_x_continuous(trans = "log10") +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
ggsave("results/figures/3-magotu_rohdes_vs_rel_cov.pdf", width = 12, height = 15)

genera_to_plot <- df_plot %>% 
        group_by(magOTU) %>% mutate(n = n_distinct(Host)) %>% filter(n > 1) %>% 
        arrange(n) %>% pull(Genus) %>% unique

df_rohdes_prev <- df_plot %>% 
        group_by(magOTU) %>% mutate(n = n_distinct(Host)) %>% filter(n > 1) %>% select(!n) %>%
        left_join(df_rohdes %>% 
            select(!Genus) %>%
            filter(type == "prev"), by = c("magOTU" = "X_old")
          )

ggplot(df_rohdes_prev,
        aes(x = factor(Host, host_order), y = prevalence_in_host,
            color = rohdes_index, group = sample)
        ) +
  geom_point(size = 4, color = "black") +
  geom_point(size = 3) +
  scale_color_gradient2(low = "#d73027", mid = "#ffffbf", high = "#1a9850", midpoint = 0.5) +
  facet_wrap(~factor(Genus, genera_to_plot), ncol = 2) +
  geom_line(aes(group = magOTU), linewidth = 1.5, alpha = 1) +
  labs(x = "Host" , color = "Rohdes index", y = "Prevalence in host") +
  make_theme(setFill = F, setCol = F, leg_pos = "right",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             modify_guide = F,
             axis_y_title = 10
  ) +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
ggsave("results/figures/3-magotu_rohdes_with_prev.pdf", width = 20, height = 25)

# using paginate_save over MAG_species_name_final

df_rohdes_prev <- df_plot %>% 
        group_by(magOTU) %>% mutate(n = n_distinct(Host)) %>% filter(n > 0) %>% select(!n) %>%
        left_join(df_rohdes %>% 
            select(!Genus) %>%
            filter(type == "prev"), by = c("magOTU" = "X_old", "MAG_species_name_final" = "MAG_species_name_final")
          )

p <- ggplot(df_rohdes_prev,
            aes(x = factor(Host, host_order), y = prevalence_in_host, color = rohdes_index, group = sample)) +
      geom_point(size = 4, color = "black") +
      geom_point(size = 3) +
      scale_color_gradient2(low = "#d73027", mid = "#ffffbf", high = "#1a9850", midpoint = 0.5) +
      facet_wrap_paginate(~MAG_species_name_final, ncol = 2, nrow = 7, page = 1) +
      geom_line(aes(group = magOTU), linewidth = 2, alpha = 1) +
      labs(x = "Host" , color = "Rohdes index", y = "Prevalence in host") +
      make_theme(setFill = F, setCol = F, leg_pos = "right",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             modify_guide = F,
             axis_y_title = 10
  ) +
  ylim(0,1) +
  theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )

paginate_save(p, "MAG_species_name_final", "results/figures/3-magotu_rohdes/3-magotu_rohdes_with_prev.pdf", pass_nrow = 6, pass_ncol = 2)


df_rohdes_prev <- df_plot %>% 
        group_by(magOTU) %>% mutate(n = n_distinct(Host)) %>% filter(n > 1) %>% select(!n) %>%
        left_join(df_rohdes %>% 
            select(!Genus) %>%
            filter(type == "prev"), by = c("magOTU" = "X_old", "MAG_species_name_final" = "MAG_species_name_final")
          )

p <- ggplot(df_rohdes_prev,
            aes(x = factor(Host, host_order), y = prevalence_in_host, color = rohdes_index, group = sample)) +
      geom_point(size = 4, color = "black") +
      geom_point(size = 3) +
      scale_color_gradient2(low = "#d73027", mid = "#ffffbf", high = "#1a9850", midpoint = 0.5) +
      facet_wrap_paginate(~MAG_species_name_final, ncol = 2, nrow = 7, page = 1) +
      geom_line(aes(group = magOTU), linewidth = 2, alpha = 1) +
      labs(x = "Host" , color = "Rohdes index", y = "Prevalence in host") +
      make_theme(setFill = F, setCol = F, leg_pos = "right",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             modify_guide = F,
             axis_y_title = 10
  ) +
  ylim(0,1) +
  theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )

paginate_save(p, "MAG_species_name_final", "results/figures/3-magotu_rohdes/3-magotu_rohdes_with_prev_2_shared.pdf", pass_nrow = 6, pass_ncol = 2)



# How are the specialist and generaliist species related.  Do we see host specificity at the sub-species level for the generalist (did you try the SNP calling/instrrain approach?)

# also, we really need to do the phylogenies and formally test for co-diversification. 
# this will go hand in hand with specificity index. 
# the non-specific species may still cluster by host at the sub-species level.

```

# Test for correlation between reads and coverage

```{r}
df_rowsums = data.frame(rowSums(abundance_df_matrix_pa_all)) %>%
              rownames_to_column("Sample")
colnames(df_rowsums) = c("Sample", "rowsum")
df_rowsums_cov = data.frame(rowSums(abundance_df_matrix_all)) %>%
              rownames_to_column("Sample")
colnames(df_rowsums_cov) = c("Sample", "rowsum_cov")
plot_corr_species_num <- df_reads %>%
                          select(Sample, MAGs_DB_mf) %>%
                            left_join(df_rowsums) %>%
                            left_join(df_rowsums_cov) %>%
                            mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
                            mutate(Location = Vectorize(get_location_from_sample_name)(Sample))

ggplot(plot_corr_species_num, aes(x = MAGs_DB_mf, y = rowsum)) +
  geom_point() +
  geom_smooth(color = "#000000", method = "lm", se = FALSE) +
  stat_cor(digits = 3, size = 5) +
  scale_color_manual(values=host_order_color) +
  scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  labs(x = "Number of reads mapping to MAGs", y = "Number of magOTUs detected") +
  make_theme(leg_pos = "bottom", setCol = F,
             axis_x_title = 16, axis_y_title = 16,
             guide_nrow = 1)
  ggsave("results/figures/09-Correlation_species_num_reads.pdf")


ggplot(plot_corr_species_num, aes(x = MAGs_DB_mf, y = rowsum, color = Host)) +
  stat_cor(digits = 3, size = 5) +
  geom_smooth(method = "lm", se = FALSE) +
  geom_point(aes(shape = Location), size = 3) +
  scale_color_manual(values=host_order_color_dark) +
  scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  labs(x = "Number of reads mapping to MAGs", y = "Number of magOTUs detected") +
  make_theme(leg_pos = "bottom", setCol = F,
             axis_x_title = 16, axis_y_title = 16,
             guide_nrow = 3)
  ggsave("results/figures/09-Correlation_species_num_reads_by_host.pdf")


ggplot(plot_corr_species_num, aes(x = MAGs_DB_mf, y = rowsum_cov, color = Host)) +
  stat_cor(digits = 3, size = 5) +
  geom_smooth(method = "lm", se = FALSE) +
  geom_point(aes(shape = Location), size = 3) +
  scale_color_manual(values=host_order_color_dark) +
  scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  labs(x = "Number of reads mapping to MAGs", y = "Sum of coverage of all magOTUs") +
  make_theme(leg_pos = "bottom", setCol = F,
             axis_x_title = 16, axis_y_title = 16,
             guide_nrow = 3)
  

# g <- ggplot(plot_corr_species_num, aes(x = MAGs_DB, y = rowsum, color = Host)) +
#   stat_cor(digits = 3, size = 5) +
#   geom_smooth(method = "lm", se = FALSE) +
#   geom_point(aes(shape = Location, name = Sample), size = 3) +
#   scale_color_manual(values=host_order_color_dark) +
#   scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
#   labs(x = "Number of reads mapping to MAGs", y = "Number of magOTUs detected") +
#   make_theme(leg_pos = "bottom", setCol = F,
#              axis_x_title = 16, axis_y_title = 16,
#              guide_nrow = 3)
# ggplotly(g)
```

# plot presence of magOTUs and genera


# Overall heatmaps

```{r plots}
df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix_pa %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)


row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0, bar_width = 0.9),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_pa), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color),
                              border = T
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_pa) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
pdf("results/figures/09-presence_absence_heatmap.pdf", width = 10, height = 15)

column_split <- rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_pa) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_pa),
        col = c("#e1ebf4", "#081d58"),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Presence"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()

pdf("results/figures/09-presence_absence_heatmap_cluster.pdf", width = 10, height = 15)

heatmap_obj = Heatmap(t(abundance_df_matrix_pa),
        col = c("#e1ebf4", "#081d58"),
        clustering_method_columns = "ward.D2",
        clustering_method_rows = "ward.D2",
        # column_split = column_split,
        # row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Presence"),
        # top_annotation = column_ba, 
        # right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
        cluster_columns = T,
          cluster_rows = T)
draw(heatmap_obj, merge_legend = TRUE)

for (genus in genera) {
  pdf(paste("results/figures/09-presence_absence_heatmap_cluster_sep/", genus, "_heatmap.pdf"), width = 10, height = 15)
  # genus = "g__Lactobacillus"
  if (dim(abundance_df_matrix_pa %>% select(matches(genus)))[[2]] == 0) {
    next
  }
  abundance_df_matrix_plot <- abundance_df_matrix_pa %>% 
                                select(matches(genus)) %>%
                                rownames_to_column("sample") %>%
                                pivot_longer(cols = -sample, names_to = "magOTU", values_to = "presence") %>%
                                left_join(df_plot %>% select(MAG_species_name_final, magOTU) %>% unique, by = c("magOTU" = "magOTU")) %>%
                                select(-magOTU) %>%
                                pivot_wider(names_from = MAG_species_name_final, values_from = presence) %>%
                                column_to_rownames("sample") %>%
                                as.matrix
  if (dim(abundance_df_matrix_plot)[2] == 0) {
    next
  }
  row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
  column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
  row_split <- colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
  heatmap_obj = Heatmap(t(abundance_df_matrix_plot),
          col = c("#e1ebf4", "#081d58"),
          clustering_method_columns = "ward.D2",
          clustering_method_rows = "ward.D2",
          # column_split = column_split,
          row_split = row_split,
          column_dend_side = "bottom",
          column_title_gp = grid::gpar(fontsize = 0),
          row_title_gp = grid::gpar(fontsize = 0),
          border = T,
          heatmap_legend_param = list(title = "Presence"),
          # top_annotation = column_ba, 
          # right_annotation = row_ra,
          left_annotation = row_la,
          bottom_annotation = column_ha,
          column_names_gp = grid::gpar(fontsize = 0),
          row_names_gp = grid::gpar(fontsize = 0),
          cluster_row_slices = F,
          cluster_column_slices = F,
          cluster_columns = T,
            cluster_rows = T)
  draw(heatmap_obj, merge_legend = TRUE, column_title = genus)
  dev.off()
}

for (genus in genera) {
  # genus = "g__Lactobacillus"
  if (dim(abundance_df_matrix_pa %>% select(matches(genus)))[[2]] == 0) {
    next
  }
  abundance_df_matrix_plot <- abundance_df_matrix_pa %>% 
                                select(matches(genus)) %>%
                                rownames_to_column("sample") %>%
                                pivot_longer(cols = -sample, names_to = "magOTU", values_to = "presence") %>%
                                left_join(df_plot %>% select(MAG_species_name_final, magOTU) %>% unique, by = c("magOTU" = "magOTU")) %>%
                                select(-magOTU) %>%
                                pivot_wider(names_from = MAG_species_name_final, values_from = presence) %>%
                                column_to_rownames("sample") %>%
                                as.matrix
  if (dim(abundance_df_matrix_plot)[2] == 0) {
    next
  }
  row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
  column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
  column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_plot), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color),
                              border = T
                              )
  row_split <- colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus) %>% as.factor
  heatmap_obj = Heatmap(t(abundance_df_matrix_plot),
          col = c("#e1ebf4", "#081d58"),
          clustering_method_columns = "ward.D2",
          clustering_method_rows = "ward.D2",
          # column_split = column_split,
          row_split = row_split,
          column_dend_side = "bottom",
          column_title_gp = grid::gpar(fontsize = 0),
          row_title_gp = grid::gpar(fontsize = 0),
          border = T,
          heatmap_legend_param = list(title = "Presence"),
          # top_annotation = column_ba, 
          # right_annotation = row_ra,
          left_annotation = row_la,
          bottom_annotation = column_ha,
          top_annotation = column_ba,
          column_names_gp = grid::gpar(fontsize = 0),
          row_names_gp = grid::gpar(fontsize = 10),
          cluster_row_slices = F,
          cluster_column_slices = F,
          cluster_columns = T,
            cluster_rows = T)
  pdf(paste0("results/figures/09-presence_absence_heatmap_cluster_sep/", genus, "_heatmap.pdf"), width = 10, height = 15)
  draw(heatmap_obj, merge_legend = TRUE, column_title = genus)
  dev.off()
}

for (genus in genera) {
  # genus = "g__Lactobacillus"
  if (dim(abundance_df_matrix_pa_all %>% select(matches(genus)))[[2]] == 0) {
    next
  }
  abundance_df_matrix_plot <- abundance_df_matrix_pa_all %>% 
                                filter(!rownames(.) %in% samples_to_exclude) %>%
                                select(matches(genus)) %>%
                                rownames_to_column("sample") %>%
                                pivot_longer(cols = -sample, names_to = "magOTU", values_to = "presence") %>%
                                left_join(df_plot %>% select(MAG_species_name_final, magOTU) %>% unique, by = c("magOTU" = "magOTU")) %>%
                                select(-magOTU) %>%
                                pivot_wider(names_from = MAG_species_name_final, values_from = presence) %>%
                                column_to_rownames("sample") %>%
                                as.matrix
  if (dim(abundance_df_matrix_plot)[2] == 0) {
    next
  }
  row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
  column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
  column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_plot), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color),
                              border = T
                              )
  row_split <- colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus) %>% as.factor
  col_split <- rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location) %>% as.factor
  heatmap_obj = Heatmap(t(abundance_df_matrix_plot),
          col = c("#e1ebf4", "#081d58"),
          clustering_method_columns = "ward.D2",
          clustering_method_rows = "ward.D2",
          column_split = col_split,
          row_split = row_split,
          column_dend_side = "bottom",
          column_title_gp = grid::gpar(fontsize = 0),
          row_title_gp = grid::gpar(fontsize = 10),
          border = T,
          heatmap_legend_param = list(title = "Presence"),
          # top_annotation = column_ba, 
          # right_annotation = row_ra,
          left_annotation = row_la,
          bottom_annotation = column_ha,
          top_annotation = column_ba,
          column_names_gp = grid::gpar(fontsize = 0),
          row_names_gp = grid::gpar(fontsize = 10),
          cluster_row_slices = F,
          cluster_column_slices = F,
          cluster_columns = T,
            cluster_rows = T)
  pdf(paste0("results/figures/09-presence_absence_heatmap_cluster_sep/", genus, "_location_host_clustering_heatmap.pdf"), width = 10, height = 15)
  draw(heatmap_obj, merge_legend = TRUE, column_title = genus)
  dev.off()
}

for (genus in genera) {
  for (host in host_order) {
    # genus = "g__Lactobacillus"
    # host = host_order[[1]]
    if (dim(abundance_df_matrix_pa_all %>% select(matches(genus)))[[2]] == 0) {
      next
    }
    locations_covered = 0
    locations_covered = rownames(abundance_df_matrix_all) %>% as.data.frame() %>% 
                          mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% 
                          mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>%
                          filter(Host == host) %>% pull(Location) %>% unique
    if (locations_covered %>% length == 1) {
      next
    }
    abundance_df_matrix_plot <- abundance_df_matrix_pa_all %>% 
                                  filter(!rownames(.) %in% samples_to_exclude) %>%
                                  select(matches(genus)) %>%
                                  rownames_to_column("sample") %>%
                                  mutate(Host = Vectorize(get_host_from_sample_name)(sample)) %>%
                                  filter(Host == host) %>%
                                  select(-Host) %>%
                                  pivot_longer(cols = -sample, names_to = "magOTU", values_to = "presence") %>%
                                  left_join(df_plot %>% select(MAG_species_name_final, magOTU) %>% unique, by = c("magOTU" = "magOTU")) %>%
                                  select(-magOTU) %>%
                                  pivot_wider(names_from = MAG_species_name_final, values_from = presence) %>%
                                  column_to_rownames("sample") %>%
                                  as.matrix
    if (dim(abundance_df_matrix_plot)[2] == 0) {
      next
    }
    if (sum(colSums(abundance_df_matrix_plot) > 0) == 0) {
      next
    }
    # remove columns with row sum 0
    if (sum(colSums(abundance_df_matrix_plot) > 0) == 1) {
      name_col = colnames(abundance_df_matrix_plot)[colSums(abundance_df_matrix_plot) > 0]
      abundance_df_matrix_plot = abundance_df_matrix_plot[, colSums(abundance_df_matrix_plot) > 0] %>% as.matrix
      colnames(abundance_df_matrix_plot) = name_col
    } else {
      abundance_df_matrix_plot <- abundance_df_matrix_plot[, colSums(abundance_df_matrix_plot) > 0]
    }
    row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus),
                        col = list(Genus = unlist(genusColors)), border = T
                      )
    column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                                col = list(Host = host_order_color
                                ), border = T
                            )
    column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                                `Number reads (MAGs)` = anno_barplot(rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_reads %>% select(Sample, MAGs_DB_mf) %>% unique, by = c(. = "Sample")) %>% pull(MAGs_DB_mf) %>% as.numeric,
                                  gp = gpar(fill = "#081d58", color = "#081d58"),
                                  baseline = 0),
                                 Colony = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Colony = Vectorize(get_colony_from_sample_name)(.)) %>% pull(Colony),
                                col = list(Location = location_order_color,
                                           Colony = make_col_list(rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Colony = Vectorize(get_colony_from_sample_name)(.)) %>% pull(Colony) %>% as.vector %>% unique)
                                ),
                                border = T
                                )
    row_split <- colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus) %>% as.factor
    col_split <- rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location) %>% as.factor
    heatmap_obj = Heatmap(t(abundance_df_matrix_plot),
            col = c("#e1ebf4", "#081d58"),
            clustering_method_columns = "ward.D2",
            clustering_method_rows = "ward.D2",
            column_split = col_split,
            row_split = row_split,
            column_dend_side = "bottom",
            column_title_gp = grid::gpar(fontsize = 0),
            row_title_gp = grid::gpar(fontsize = 10),
            border = T,
            heatmap_legend_param = list(title = "Presence"),
            # top_annotation = column_ba, 
            # right_annotation = row_ra,
            left_annotation = row_la,
            bottom_annotation = column_ha,
            top_annotation = column_ba,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 10),
            cluster_row_slices = F,
            cluster_column_slices = T,
            cluster_columns = T,
              cluster_rows = T)
    pdf(paste0("results/figures/09-presence_absence_heatmap_cluster_sep/", genus, "-", host,"_location_heatmap.pdf"), width = 10, height = 15)
    draw(heatmap_obj, merge_legend = TRUE, column_title = paste0(genus, " in -", host))
    dev.off()
  }
}


# repeat using relative abundance

system("mkdir results/figures/09-presence_absence_heatmap_cluster_sep_rel")

for (genus in genera) {
  # genus = "g__Lactobacillus"
  if (dim(abundance_df_matrix_rel %>% select(matches(genus)))[[2]] == 0) {
    next
  }
  abundance_df_matrix_plot <- abundance_df_matrix_rel %>% 
                                select(matches(genus)) %>%
                                rownames_to_column("sample") %>%
                                pivot_longer(cols = -sample, names_to = "magOTU", values_to = "presence") %>%
                                left_join(df_plot %>% select(MAG_species_name_final, magOTU) %>% unique, by = c("magOTU" = "magOTU")) %>%
                                select(-magOTU) %>%
                                pivot_wider(names_from = MAG_species_name_final, values_from = presence) %>%
                                column_to_rownames("sample") %>%
                                as.matrix
  if (dim(abundance_df_matrix_plot)[2] == 0) {
    next
  }
  row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
  column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
  column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_plot), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color),
                              border = T
                              )
  row_split <- colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus) %>% as.factor
  heatmap_obj = Heatmap(t(abundance_df_matrix_plot),
          col = c("#e1ebf4", "#081d58"),
          clustering_method_columns = "ward.D2",
          clustering_method_rows = "ward.D2",
          # column_split = column_split,
          row_split = row_split,
          column_dend_side = "bottom",
          column_title_gp = grid::gpar(fontsize = 0),
          row_title_gp = grid::gpar(fontsize = 0),
          border = T,
          heatmap_legend_param = list(title = "Presence"),
          # top_annotation = column_ba, 
          # right_annotation = row_ra,
          left_annotation = row_la,
          bottom_annotation = column_ha,
          top_annotation = column_ba,
          column_names_gp = grid::gpar(fontsize = 0),
          row_names_gp = grid::gpar(fontsize = 10),
          cluster_row_slices = F,
          cluster_column_slices = F,
          cluster_columns = T,
            cluster_rows = T)
  pdf(paste0("results/figures/09-presence_absence_heatmap_cluster_sep_rel/", genus, "_heatmap.pdf"), width = 10, height = 15)
  draw(heatmap_obj, merge_legend = TRUE, column_title = genus)
  dev.off()
}

for (genus in genera) {
  # genus = "g__Lactobacillus"
  if (dim(abundance_df_matrix_rel_all %>% select(matches(genus)))[[2]] == 0) {
    next
  }
  abundance_df_matrix_plot <- abundance_df_matrix_rel_all %>% 
                                filter(!rownames(.) %in% samples_to_exclude) %>%
                                select(matches(genus)) %>%
                                rownames_to_column("sample") %>%
                                pivot_longer(cols = -sample, names_to = "magOTU", values_to = "presence") %>%
                                left_join(df_plot %>% select(MAG_species_name_final, magOTU) %>% unique, by = c("magOTU" = "magOTU")) %>%
                                select(-magOTU) %>%
                                pivot_wider(names_from = MAG_species_name_final, values_from = presence) %>%
                                column_to_rownames("sample") %>%
                                as.matrix
  if (dim(abundance_df_matrix_plot)[2] == 0) {
    next
  }
  row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
  column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
  column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_plot), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color),
                              border = T
                              )
  row_split <- colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus) %>% as.factor
  col_split <- rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location) %>% as.factor
  heatmap_obj = Heatmap(t(abundance_df_matrix_plot),
          col = c("#e1ebf4", "#081d58"),
          clustering_method_columns = "ward.D2",
          clustering_method_rows = "ward.D2",
          column_split = col_split,
          row_split = row_split,
          column_dend_side = "bottom",
          column_title_gp = grid::gpar(fontsize = 0),
          row_title_gp = grid::gpar(fontsize = 10),
          border = T,
          heatmap_legend_param = list(title = "Presence"),
          # top_annotation = column_ba, 
          # right_annotation = row_ra,
          left_annotation = row_la,
          bottom_annotation = column_ha,
          top_annotation = column_ba,
          column_names_gp = grid::gpar(fontsize = 0),
          row_names_gp = grid::gpar(fontsize = 10),
          cluster_row_slices = F,
          cluster_column_slices = F,
          cluster_columns = T,
            cluster_rows = T)
  pdf(paste0("results/figures/09-presence_absence_heatmap_cluster_sep_rel/", genus, "_location_host_clustering_heatmap.pdf"), width = 10, height = 15)
  draw(heatmap_obj, merge_legend = TRUE, column_title = genus)
  dev.off()
}

for (genus in genera) {
  for (host in host_order) {
    # genus = "g__Lactobacillus"
    # host = host_order[[1]]
    if (dim(abundance_df_matrix_rel_all %>% select(matches(genus)))[[2]] == 0) {
      next
    }
    locations_covered = 0
    locations_covered = rownames(abundance_df_matrix_all) %>% as.data.frame() %>% 
                          mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% 
                          mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>%
                          filter(Host == host) %>% pull(Location) %>% unique
    if (locations_covered %>% length == 1) {
      next
    }
    abundance_df_matrix_plot <- abundance_df_matrix_rel_all %>% 
                                  filter(!rownames(.) %in% samples_to_exclude) %>%
                                  select(matches(genus)) %>%
                                  rownames_to_column("sample") %>%
                                  mutate(Host = Vectorize(get_host_from_sample_name)(sample)) %>%
                                  filter(Host == host) %>%
                                  select(-Host) %>%
                                  pivot_longer(cols = -sample, names_to = "magOTU", values_to = "presence") %>%
                                  left_join(df_plot %>% select(MAG_species_name_final, magOTU) %>% unique, by = c("magOTU" = "magOTU")) %>%
                                  select(-magOTU) %>%
                                  pivot_wider(names_from = MAG_species_name_final, values_from = presence) %>%
                                  column_to_rownames("sample") %>%
                                  as.matrix
    if (dim(abundance_df_matrix_plot)[2] == 0) {
      next
    }
    if (sum(colSums(abundance_df_matrix_plot) > 0) == 0) {
      next
    }
    # remove columns with row sum 0
    if (sum(colSums(abundance_df_matrix_plot) > 0) == 1) {
      name_col = colnames(abundance_df_matrix_plot)[colSums(abundance_df_matrix_plot) > 0]
      abundance_df_matrix_plot = abundance_df_matrix_plot[, colSums(abundance_df_matrix_plot) > 0] %>% as.matrix
      colnames(abundance_df_matrix_plot) = name_col
    } else {
      abundance_df_matrix_plot <- abundance_df_matrix_plot[, colSums(abundance_df_matrix_plot) > 0]
    }
    row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus),
                        col = list(Genus = unlist(genusColors)), border = T
                      )
    column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                                col = list(Host = host_order_color
                                ), border = T
                            )
    column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                                `Number reads (MAGs)` = anno_barplot(rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_reads %>% select(Sample, MAGs_DB_mf) %>% unique, by = c(. = "Sample")) %>% pull(MAGs_DB_mf) %>% as.numeric,
                                  gp = gpar(fill = "#081d58", color = "#081d58"),
                                  baseline = 0),
                                 Colony = rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Colony = Vectorize(get_colony_from_sample_name)(.)) %>% pull(Colony),
                                col = list(Location = location_order_color,
                                           Colony = make_col_list(rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Colony = Vectorize(get_colony_from_sample_name)(.)) %>% pull(Colony) %>% as.vector %>% unique)
                                ),
                                border = T
                                )
    row_split <- colnames(abundance_df_matrix_plot) %>% as.data.frame() %>% left_join(df_plot %>% select(MAG_species_name_final, Genus) %>% unique, by = c(. = "MAG_species_name_final")) %>% pull(Genus) %>% as.factor
    col_split <- rownames(abundance_df_matrix_plot) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location) %>% as.factor
    heatmap_obj = Heatmap(t(abundance_df_matrix_plot),
            col = c("#e1ebf4", "#081d58"),
            clustering_method_columns = "ward.D2",
            clustering_method_rows = "ward.D2",
            column_split = col_split,
            row_split = row_split,
            column_dend_side = "bottom",
            column_title_gp = grid::gpar(fontsize = 0),
            row_title_gp = grid::gpar(fontsize = 10),
            border = T,
            heatmap_legend_param = list(title = "Presence"),
            # top_annotation = column_ba, 
            # right_annotation = row_ra,
            left_annotation = row_la,
            bottom_annotation = column_ha,
            top_annotation = column_ba,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 10),
            cluster_row_slices = F,
            cluster_column_slices = T,
            cluster_columns = T,
              cluster_rows = T)
    pdf(paste0("results/figures/09-presence_absence_heatmap_cluster_sep_rel/", genus, "-", host,"_location_heatmap.pdf"), width = 10, height = 15)
    draw(heatmap_obj, merge_legend = TRUE, column_title = paste0(genus, " in -", host))
    dev.off()
  }
}

system("rename 's/\\s/_/g' results/figures/09-presence_absence_heatmap_cluster_sep_rel/*")
system("rename 's/\\s/_/g' results/figures/09-presence_absence_heatmap_cluster_sep/*")

df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix_copies %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)
df_host_nums[is.na(df_host_nums)] <- 0
row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_copies) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
# make it 1 and 0 from numbers before counting
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_copies) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(mutate_all(abundance_df_matrix_copies, function(x) ifelse(x > 0, 1, 0))), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color)
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_copies) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
pdf("results/figures/09-coverage_heatmap_copies.pdf")

column_split <- rownames(abundance_df_matrix_copies) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_copies) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_copies),
        col = colorRamp2(c(0, 1e2, 1e4, 1e5, 1e6, 1e8), colors = c(brewer.pal(9, "Blues")[1], brewer.pal(9, "Blues")[2], brewer.pal(9, "Blues")[3], brewer.pal(9, "Blues")[4], brewer.pal(9, "Blues")[7], brewer.pal(9, "Blues")[8])),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Coverage"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()

df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix_rel %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)
df_host_nums[is.na(df_host_nums)] <- 0
row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_rel) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_rel) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              # `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_rel), 
                              #   gp = gpar(fill = "#081d58", color = "#081d58"),
                              #   baseline = 0),
                              col = list(Location = location_order_color)
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_rel) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
pdf("results/figures/09-figures/09-coverage_heatmap_relative.pdf")

column_split <- rownames(abundance_df_matrix_rel) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_rel) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_rel),
        col = colorRamp2(c(0, 1e-4, 1e-3, 1e-2, 1e-1, 1), colors = c(brewer.pal(9, "Blues")[1], brewer.pal(9, "Blues")[2], brewer.pal(9, "Blues")[3], brewer.pal(9, "Blues")[4], brewer.pal(9, "Blues")[7], brewer.pal(9, "Blues")[8])),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Coverage"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()

g_list <- list()
for (a_sample in unique(df_plot$sample)) {
  g <- ggplot(df_plot %>% filter(sample == a_sample), aes(x = prevalence_in_host, 
                    y = coverage)
                ) +
                geom_point(aes(color = Genus), size = 5) +
                geom_smooth(method = "lm", se = FALSE, color = "#000000") +
                ggtitle(a_sample) +
                stat_cor(digits = 3, size = 5) +
                scale_y_continuous(trans = "log10") +
                scale_color_manual(values=genusColors) +
                labs(x = "Prevalence in host", y = "Coverage") +
                make_theme(leg_pos = "none", setCol = F,
                           axis_x_title = 10, axis_y_title = 10,
                           guide_nrow = 1)
  g_list[[a_sample]] <- g
}
pdf("results/figures/09-abundance_vs_prevalence_in_host.pdf", width = 10, height = 15)
marrangeGrob(g_list, nrow = 5, ncol = 2)
dev.off()

# genus-level heatmap
# x axis has genus and y host species
# color is prevalence in host
df_plot %>% select(sample, Host) %>% unique() %>% pull(Host) %>% table
df_genus_prev_map <- df_plot %>%
                      mutate(detected = ifelse(coverage > 1, 1, 0)) %>%
                      filter(detected == 1) %>%
                      filter(!is.na(Genus)) %>%
                      ungroup() %>%
                        group_by(Host, Genus) %>%
                        # num samples in which the species/genus is detected
                          summarise(Host, Genus, number = n_distinct(sample)) %>%
                          unique() %>%
                              ungroup() %>%
                              mutate(host_num_samples = ifelse(Host == "Apis mellifera", 34, NA)) %>%
                              mutate(host_num_samples = ifelse(Host == "Apis cerana", 45, host_num_samples)) %>%
                              mutate(host_num_samples = ifelse(Host == "Apis dorsata", 44, host_num_samples)) %>%
                              mutate(host_num_samples = ifelse(Host == "Apis florea", 42, host_num_samples)) %>%
                              mutate(host_num_samples = ifelse(Host == "Apis andreniformis", 30, host_num_samples)) %>%
                              mutate(genus_prev = number/host_num_samples) %>%
                              ungroup() %>%
                              group_by(Genus) %>%
                              mutate(min_genus_prev = min(genus_prev)) %>%
                              mutate(max_genus_prev = max(genus_prev)) %>%
                              ungroup() %>%
                              select(Host, Genus, genus_prev) %>%
                              mutate(Genus = Vectorize(clean_g_name)(Genus))
                                # pivot_wider(names_from = Host, values_from = genus_prev) %>%
                                #   column_to_rownames("Genus") %>%
                                #   # replace NA with 0
                                #   replace(is.na(.), 0) %>%
                                #     as.matrix
# df_genus_prev_map %>% view
df_plot %>%
                      mutate(detected = ifelse(coverage > 1, 1, 0)) %>%
                      filter(detected == 1) %>%
                      filter(!is.na(Genus)) %>%
                      ungroup() %>%
                        group_by(Host, Genus) %>%
                        # num samples in which the species/genus is detected
                          summarise(Host, Genus, number = n_distinct(sample)) %>%
                          unique() %>%
                              ungroup() %>%
                              mutate(host_num_samples = ifelse(Host == "Apis mellifera", 34, NA)) %>%
                              mutate(host_num_samples = ifelse(Host == "Apis cerana", 45, host_num_samples)) %>%
                              mutate(host_num_samples = ifelse(Host == "Apis dorsata", 44, host_num_samples)) %>%
                              mutate(host_num_samples = ifelse(Host == "Apis florea", 42, host_num_samples)) %>%
                              mutate(host_num_samples = ifelse(Host == "Apis andreniformis", 30, host_num_samples)) %>%
                              mutate(genus_prev = number/host_num_samples) %>%
                              ungroup() %>%
                              group_by(Genus) %>%
                              mutate(min_genus_prev = min(genus_prev)) %>%
                              mutate(max_genus_prev = max(genus_prev)) %>% view
                              filter(genus_prev > 0.1) %>% 
                              pull(Genus) %>% unique() %>% length
# df_genus_prev_map %>% view
genus_order_spl <- c(
    "Lactobacillus",
    "Bifidobacterium",
    "Bombilactobacillus",
    "Snodgrassella",
    "Gilliamella",
    "Bartonella_A",
    "Frischella",
    "Apibacter",
    "Commensalibacter",
    "Apilactobacillus",
    "CALYQJ01",
    "CALYQQ01",
    # "Bombella",
    "Dysgonomonas",
    "Saezia",
    "Entomomonas",
    "Pectinatus",
    "Pluralibacter",
    "Parolsenella",
    "Enterobacter",
    "WRHT01"
    # "Spiroplasma",
    # "Zymobacter",
    # "Melissococcus",
    # "Cronobacter",
    # "KACC-21507",
    # "Rosenbergiella"
  )   
ggplot(df_genus_prev_map) +
  geom_tile(aes(x = factor(Genus, genus_order_spl), y = factor(Host, rev(host_order)), fill = genus_prev)) +
  labs(x = "Host species", y = "Genus", fill = "Prevalence in host") +
  scale_fill_gradient(low = "#ffffff", high = "#0570b0") +
  coord_equal() +
  make_theme(leg_pos = "right", setCol = F, setFill = F,
             axis_x_title = 16, axis_y_title = 16,
             x_angle = 90, x_hj = 1, x_vj = 1,
             y_hj = 1, modify_guide = F,
             guide_nrow = 1)
  ggsave("results/figures/09-figures/09-genus_prevalence_heatmap.pdf")
```

# Community measures

Do not do this on the front-end of Curnagl. Run elsewhere or submit as a job!

```{r}
df_plot_cum_curve_by_loc <- read.csv("results/figures/magotu_cum_curves_loc_all.csv")

df_plot_cum_curve <- read.csv("results/figures/magotu_cum_curves.csv") %>%
                      mutate(host = paste0("Apis ", host))

ggplot(data = df_plot_cum_curve, aes(x = sample_size, y = num_magotus, color = factor(host, host_order))) +
                      geom_jitter(position = position_dodge(width=0.7)) +
                        geom_smooth(method = "loess", se = F) +
                          labs(color = "Host species", x = "Number of Bees", y = "Number of magOTUs") +
                          scale_color_manual(values=host_order_color_dark) +
                            make_theme(leg_pos = "bottom", setCol = F,
                                       axis_x_title = 16, axis_y_title = 16,
                                       guide_nrow = 1)
                            ggsave("results/figures/09-Cumulative_curve.pdf")

host_order_color_ext <- c("Apis mellifera (India)" = "#4292c6",
                          "Apis mellifera (Malaysia)" = "#08519c",
                          "Apis mellifera (Japan)" = "#a6cee3",
                          "Apis mellifera (Switzerland)" = "#41b6c4",
                          "Apis cerana (Japan)" = "#fbb4ae", 
                          "Apis cerana (India)" = "#fb9a99", 
                          "Apis cerana (Malaysia)" = "#e31a1c", 
                          "Apis dorsata (Malaysia)" = "#984ea3",
                          "Apis dorsata (India)" = "#cab2d6",
                          "Apis florea (Malaysia)" = "#4daf4a",
                          "Apis florea (India)" = "#b2df8a",
                          "Apis andreniformis (Malaysia)" = "#ff7f00")
ggplot(data = df_plot_cum_curve_by_loc, aes(x = sample_size, y = num_magotus, color = factor(host))) +
                      geom_jitter(position = position_dodge(width=0.7)) +
                        geom_smooth(method = "loess", se = T) +
                          labs(color = "Host species", x = "Number of Bees", y = "Number of magOTUs") +
                          scale_color_manual(values=host_order_color_ext) +
                            make_theme(leg_pos = "bottom", leg_size = 11,
                                       axis_x_title = 16, axis_y_title = 16,
                                       setCol = F, guide_nrow = 4) +
                              theme(legend.spacing.x = unit(1.0, 'cm'))
                            ggsave("results/figures/09-Cumulative_curve_by_location.pdf")

# species richness boxplot

df_box_plot <- df_plot %>%
  group_by(sample) %>%
  summarise(num_magotus = n_distinct(magOTU)) %>%
  # unique %>%
    mutate(Host = Vectorize(get_host_from_sample_name)(sample))

ggplot(df_box_plot) +
  geom_boxplot(aes(x = factor(Host, host_order), y = num_magotus, fill = factor(Host, host_order)), outlier.shape = NA) +
  geom_jitter(aes(x = factor(Host, host_order), y = num_magotus), width = 0.1, size = 2) +
  labs(x = "Host species", y = "Number of bacterial species") +
  scale_fill_manual(values=host_order_color_dark) +
  geom_signif(aes(x = factor(Host, host_order), y = num_magotus),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                 c("Apis mellifera", "Apis cerana"),
                                #  c("Apis mellifera", "Apis florea"),
                                 c("Apis mellifera", "Apis dorsata"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis cerana", "Apis florea"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis cerana", "Apis andreniformis")),
                map_signif_level = TRUE, textsize = 6, vjust = 0.5) +
  make_theme(leg_pos = "none", setCol = F, setFill = F,
             axis_x_title = 16, axis_y_title = 16,
             guide_nrow = 1)
  ggsave("results/figures/09-Species_per_sample_boxplot.pdf")

shared_magotus <- df_plot %>% group_by(magOTU) %>% filter(n_distinct(Host) > 1) %>% pull(magOTU) %>% unique

shared_genera <- 
  df_plot %>% 
    filter(!is.na(Genus)) %>% group_by(Genus) %>% filter(n_distinct(Host) == 5) %>% pull(Genus) %>% unique %>% as.character

# df_plot %>% 
#     filter(!is.na(Genus)) %>% group_by(Genus) %>% filter(n_distinct(Host) == 5) %>% 
#     filter(!is.na(prevalence_in_host)) %>% 
#     group_by(Host, Genus) %>%
#     summarise(number_samples_present = n_distinct(sample)) %>% 
#     ungroup() %>%
#     group_by(Genus) %>%
#     mutate(min = min(number_samples_present), max = max(number_samples_present)) %>% view


df_box_plot_shared <- df_plot %>%
  # filter to limit to species shared by 2 or more hosts
  # filter(magOTU %in% shared_magotus) %>%
  filter(Genus %in% shared_genera) %>%
  # filter(Genus != "g__Klebsiella") %>%
  group_by(sample) %>% 
  summarise(num_magotus = n_distinct(magOTU)) %>%
    mutate(Host = Vectorize(get_host_from_sample_name)(sample))
host_order_spl <- c("Apis dorsata",
                    "Apis mellifera",
                    "Apis andreniformis",
                    "Apis cerana",
                    "Apis florea"
)
ggplot(df_box_plot_shared) +
  geom_boxplot(aes(x = factor(Host, host_order_spl), y = num_magotus, fill = factor(Host, host_order_spl)), outlier.shape = NA) +
  geom_jitter(aes(x = factor(Host, host_order_spl), y = num_magotus), width = 0.1, size = 2) +
  labs(x = "Host species", y = "Number of bacterial species") +
  scale_fill_manual(values=host_order_color_dark) +
  geom_signif(aes(x = factor(Host, host_order_spl), y = num_magotus),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis mellifera", "Apis florea"),
                                 c("Apis mellifera", "Apis dorsata"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis florea"),
                                 c("Apis cerana", "Apis florea"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis cerana", "Apis andreniformis")),
                map_signif_level = TRUE, textsize = 6, vjust = 0.5) +
  ylim(0, 60) +
  make_theme(leg_pos = "none", setCol = F, setFill = F,
             axis_x_title = 16, axis_y_title = 16,
             guide_nrow = 1)
  ggsave("results/figures/09-Species_per_sample_boxplot_shared.pdf")


# code for betapart functions from
# https://github.com/cran/betapart

# Latest commit a222c6d on Feb 24, 2012
betapart.core <- function(x){
	
	# test for a binary numeric matrix or data.frame
	if(! is.matrix(x)){
		x<-as.matrix(x)
  	}
	
	if(! is.numeric(x))
    	stop("The data in x is not numeric.",call.=TRUE)
	
	# simple test for binary data
	xvals <-  unique(as.vector(x))
	if (any(!is.element(xvals, c(0, 1)))) 
        stop("The table contains values other than 0 and 1: data should be presence/absence.", call. = TRUE)

	shared <- x %*% t(x)
    not.shared <-  abs(sweep(shared, 2, diag(shared)))
		
	sumSi <- sum(diag(shared)) # species by site richness
    St <- sum(colSums(x) > 0)  # regional species richness
    a <- sumSi - St            # multi site shared species term

    sum.not.shared <- not.shared + t(not.shared)
    max.not.shared <- pmax(not.shared, t(not.shared))
    min.not.shared <- pmin(not.shared, t(not.shared))

	computations<-list(data=x, sumSi=sumSi, St=St, a=a, shared=shared, not.shared=not.shared, 
	                   sum.not.shared=sum.not.shared, max.not.shared=max.not.shared, 
	                   min.not.shared=min.not.shared)
    class(computations)<-"betapart"

	return(computations)
} 
# Latest commit 4b59d9c on Feb 24, 2012
beta.pair <- function(x, index.family="sorensen"){
	
	# test for a valid index
	index.family <- match.arg(index.family, c('jaccard','sorensen'))
	
	# test for pre-existing betapart objects
	if (! inherits(x, "betapart")){	
		x <- betapart.core(x)
	}
	
	# run the analysis given the index
	switch(index.family,
		sorensen = {
			beta.sim <- x$min.not.shared / (x$min.not.shared + x$shared)
			beta.sne <- ((x$max.not.shared - x$min.not.shared) / ((2 * x$shared) + x$sum.not.shared)) * (x$shared / (x$min.not.shared + x$shared))
            beta.sor <- x$sum.not.shared / (2 * x$shared + x$sum.not.shared)
                
            pairwise <- list(beta.sim=as.dist(beta.sim), beta.sne=as.dist(beta.sne),beta.sor=as.dist(beta.sor))},
		jaccard = {
			beta.jtu <- (2*x$min.not.shared) / ((2*x$min.not.shared) + x$shared)
	        beta.jne <- ((x$max.not.shared - x$min.not.shared) / (x$shared + x$sum.not.shared)) * (x$shared / ((2*x$min.not.shared) + x$shared))
	        beta.jac <- x$sum.not.shared / (x$shared + x$sum.not.shared)

	        pairwise <- list(beta.jtu=as.dist(beta.jtu), beta.jne=as.dist(beta.jne),beta.jac=as.dist(beta.jac))})

	return(pairwise)
}
beta.multi <- function(x, index.family="sorensen"){

	# test for a valid index
	index.family <- match.arg(index.family, c('jaccard','sorensen'))
	
	# test for pre-existing betapart objects
	if (! inherits(x, "betapart")){	
		x <- betapart.core(x)
	}

	maxbibj <- sum(x$max.not.shared[lower.tri(x$max.not.shared)])
    minbibj <- sum(x$min.not.shared[lower.tri(x$min.not.shared)])
	
	# run the analysis given the index
	switch(index.family,
		sorensen = {
            beta.sim <- minbibj / (minbibj + x$a)
            beta.sne <- (x$a / (minbibj + x$a)) * ((maxbibj - minbibj) / ((2 * x$a) + maxbibj + minbibj))
            beta.sor <- (minbibj + maxbibj) / (minbibj + maxbibj + (2 * x$a))

           	multi <- list(beta.SIM=beta.sim, beta.SNE=beta.sne,beta.SOR=beta.sor)},
		jaccard = {
            beta.jtu <- (2*minbibj) / ((2*minbibj) + x$a)
            beta.jne <- (x$a / ((2*minbibj) + x$a)) * ((maxbibj - minbibj) / ((x$a) + maxbibj + minbibj))
            beta.jac <- (minbibj + maxbibj) / (minbibj + maxbibj + x$a)

           	multi <- list(beta.JTU=beta.jtu, beta.JNE=beta.jne, beta.JAC=beta.jac)})

	return(multi)

}

df_magOTUs_vegan_pa <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0) %>%
                      filter(!rownames(.) %in% samples_to_exclude)
# df_pcoa <- dist_matrix_betapart$beta.sor
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "sorensen")
# test dispertion and adonis2
dist_obj <- dist_matrix_betapart$beta.sor
# disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
# anova(disp_res)
ado_res = adonis2(dist_obj ~ Species, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
# adonis_OmegaSq(ado_res)
pdf("results/figures/09-PCoA_sorensen_microbiome.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark) +
  annotate("text", x = 0.25,
                      y = 0.5,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]], "  (N=", 1+ado_res$Df[[3]], ")")
          )
dev.off()

ado_res = adonis2(dist_obj ~ Location, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
adonis_OmegaSq(ado_res)
pdf("results/figures/09-PCoA_sorensen_microbiome_location.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete %>% select(ID, Country), variable = "Country", color_add=T, color_list=location_order_color) +
  annotate("text", x = 0.25,
                      y = 0.5,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]], "  (N=", 1+ado_res$Df[[3]], ")")
          )
dev.off()

df_magOTUs_vegan_pa <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0) %>%
                      filter(!rownames(.) %in% samples_to_exclude) %>%
                      filter(!rownames(.) %in% samples_AA) %>%
                      filter(!rownames(.) %in% samples_AM)
# df_pcoa <- dist_matrix_betapart$beta.sor
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "sorensen")
# test dispertion and adonis2
dist_obj <- dist_matrix_betapart$beta.sor
ado_res = adonis2(dist_obj ~ Location, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
adonis_OmegaSq(ado_res)
pdf("results/figures/09-PCoA_sorensen_microbiome_location_noAndreMelli.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete %>% select(ID, Country), variable = "Country", color_add=T, color_list=location_order_color) +
  annotate("text", x = 0.25,
                      y = 0.5,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]], "  (N=", 1+ado_res$Df[[3]], ")")
          )
dev.off()

dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "jaccard")
dist_obj <- dist_matrix_betapart$beta.jac
disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
anova(disp_res)
ado_res = adonis2(dist_obj ~ Species, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
# adonis_OmegaSq(ado_res)
pdf("results/figures/09-PCoA_jaccard_microbiome.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_betapart$beta.jac, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark) +
  annotate("text", x = -0.25,
                      y = 0.5,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]], "  (N=", 1+ado_res$Df[[3]], ")")
          )
dev.off()


dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "sorensen")
dist_obj <- dist_matrix_betapart$beta.sim
disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
anova(disp_res)
ado_res = adonis2(dist_obj ~ Species, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
# adonis_OmegaSq(ado_res)
pdf("results/figures/09-PCoA_sorensen_microbiome_turnover.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_betapart$beta.sim, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark) +
  annotate("text", x = -0.1,
                      y = 0.5,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]], "  (N=", 1+ado_res$Df[[3]], ")")
          )
dev.off()

dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "jaccard")
dist_obj <- dist_matrix_betapart$beta.jtu
disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
anova(disp_res)
ado_res = adonis2(dist_obj ~ Species, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
# adonis_OmegaSq(ado_res)
pdf("results/figures/09-PCoA_jaccard_microbiome_turnover.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_betapart$beta.jtu, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark) +
  annotate("text", x = -0.1,
                      y = 0.5,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]], "  (N=", 1+ado_res$Df[[3]], ")")
          )
dev.off()

pcoa_plot(dist_matrix_betapart$beta.jac, df_meta_complete %>% select(ID, Species, Country), variable = "Species", shape_add=T, shape_var = "Country", color_add=T, color_list=host_order_color_dark)

dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "sorensen")
pdf("results/figures/09-PCoA_sorensen_microbiome_nestedness.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_betapart$beta.sne, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark)
dev.off()
dist_obj <- dist_matrix_betapart$beta.sne
disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
anova(disp_res)
ado_res = adonis2(dist_obj ~ Species + Location, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
adonis_OmegaSq(ado_res)

dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "jaccard")
pdf("results/figures/09-PCoA_jaccard_microbiome_nestedness.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_betapart$beta.jne, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark)
dev.off()
dist_obj <- dist_matrix_betapart$beta.jne
disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
anova(disp_res)
ado_res = adonis2(dist_obj ~ Species + Location, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
adonis_OmegaSq(ado_res)

dist_matrix_atch <- vegdist(abundance_df_matrix_copies, method = "robust.aitchison")
dist_obj <- dist_matrix_atch
disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
anova(disp_res)
ado_res = adonis2(dist_obj ~ Species + Location, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
adonis_OmegaSq(ado_res)
pdf("results/figures/09-PCoA_atchinson_microbiome.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_atch, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark) +
  annotate("text", x = 2.5,
                      y = 3.5,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]], "  (N=", 1+ado_res$Df[[3]], ")")
          )
dev.off()

dist_matrix_atch_rel <- vegdist(abundance_df_matrix_rel, method = "robust.aitchison")
dist_obj <- dist_matrix_atch_rel
disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
anova(disp_res)
ado_res = adonis2(dist_obj ~ Species, 
        rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>%
        mutate(Species = Vectorize(get_host_from_sample_name)(.)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(.))
)
# adonis_OmegaSq(ado_res)
pdf("results/figures/09-PCoA_atchinson_relative_ab_microbiome.pdf", width = 10, height = 15)
pcoa_plot(dist_matrix_atch_rel, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark) +
  annotate("text", x = 2.5,
                      y = 3.5,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]], "  (N=", 1+ado_res$Df[[3]], ")")
          )
dev.off()

# if betadisper is significant, (adonis ~ permanova)
# https://esajournals.onlinelibrary.wiley.com/doi/10.1890/12-2010.1

# # beta.SIM value of the turnover component, measured as Simpson dissimilarity
# # beta.SNE value of the nestedness component, measured as nestedness-resultant fraction of
# # Sorensen dissimilarity
# # beta.SOR value of the overall beta diversity, measured as Sorensen dissimilarity

# pcoa showing colony affiliation
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "sorensen")
my_colony_col_list2 <- c("M1" = "#fdae61",
                        "M2" = "#fee08b",
                        "M3" = "#ffffbf",
                        "M4" = "#e6f598",
                        "M5" = "#abdda4",
                        "M6" = "#66c2a5",
                        "M7" = "#3288bd",
                        "C1" = "#d53e4f",
                        "C2" = "#f46d43",
                        "C3" = "#fdae61",
                        "C4" = "#fee08b",
                        "C5" = "#ffffbf",
                        "C6" = "#e6f598",
                        "C7" = "#abdda4",
                        "C8" = "#66c2a5",
                        "C9" = "#3288bd",
                        "F1" = "#d53e4f",
                        "F2" = "#f46d43",
                        "F3" = "#fdae61",
                        "F4" = "#fee08b",
                        "F5" = "#ffffbf",
                        "F6" = "#e6f598",
                        "F7" = "#abdda4",
                        "F8" = "#66c2a5",
                        "F9" = "#3288bd",
                        "D1" = "#d53e4f",
                        "D2" = "#f46d43",
                        "D3" = "#fdae61",
                        "D4" = "#fee08b",
                        "D5" = "#ffffbf",
                        "D6" = "#e6f598",
                        "D7" = "#abdda4",
                        "D8" = "#66c2a5",
                        "D9" = "#3288bd",
                        "A1" = "#d53e4f",
                        "A2" = "#fdae61",
                        "A3" = "#e6f598",
                        "A4" = "#abdda4",
                        "A5" = "#66c2a5",
                        "A6" = "#3288bd"
                 )
my_colony_col_list <- c("M1" = "#08306b",
                        "M2" = "#08519c",
                        "M3" = "#2171b5",
                        "M4" = "#4292c6",
                        "M5" = "#6baed6",
                        "M6" = "#9ecae1",
                        "M7" = "#08306b",
                        "C1" = "#67000d",
                        "C2" = "#a50f15",
                        "C3" = "#cb181d",
                        "C4" = "#ef3b2c",
                        "C5" = "#fb6a4a",
                        "C6" = "#fc9272",
                        "C7" = "#67000d",
                        "C8" = "#a50f15",
                        "C9" = "#cb181d",
                        "F1" = "#238b45",
                        "F2" = "#006d2c",
                        "F3" = "#00441b",
                        "F4" = "#a1d99b",
                        "F5" = "#74c476",
                        "F6" = "#41ab5d",
                        "F7" = "#238b45",
                        "F8" = "#006d2c",
                        "F9" = "#00441b",
                        "D1" = "#6a51a3",
                        "D2" = "#54278f",
                        "D3" = "#3f007d",
                        "D4" = "#bcbddc",
                        "D5" = "#9e9ac8",
                        "D6" = "#807dba",
                        "D7" = "#6a51a3",
                        "D8" = "#54278f",
                        "D9" = "#3f007d",
                        "A1" = "#fdae6b",
                        "A2" = "#fd8d3c",
                        "A3" = "#f16913",
                        "A4" = "#d94801",
                        "A5" = "#a63603",
                        "A6" = "#7f2704"
                 )
pdf("results/figures/09-PCoA_sorensen_microbiome_w_colony.pdf", width = 10, height = 15) 
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete %>% select(ID, Colony_ID, Species), variable = "Colony_ID", color_add=T, color_list=my_colony_col_list, shape_add = T, shape_var = "Species")
dev.off()
pdf("results/figures/09-PCoA_sorensen_microbiome_w_colony_and_spec.pdf", width = 10, height = 15) 
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete %>% select(ID, Colony_ID, Species), variable = "Colony_ID", color_add=T, color_list=my_colony_col_list, shape_add = F, shape_var = "Species")
dev.off()
pdf("results/figures/09-PCoA_sorensen_microbiome_w_colony_and_country.pdf", width = 10, height = 15) 
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete %>% select(ID, Colony_ID, Country), variable = "Colony_ID", color_add=T, color_list=my_colony_col_list, shape_add = T, shape_var = "Country")
dev.off()
pdf("results/figures/09-PCoA_sorensen_microbiome_w_colony_and_country_by_host.pdf", width = 10, height = 15) 
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete %>% select(ID, Colony_ID, Country, Species), variable = "Colony_ID", color_add=T, color_list=my_colony_col_list2, shape_add = T, shape_var = "Country") +
  facet_wrap(~Species, scales = "free")
dev.off()


# Test colony effect in mellifera and cerana
df_magOTUs_vegan_pa_all <- abundance_df_matrix_pa_all %>%
                      filter(rowSums(.) > 0) %>%
                      filter(rownames(.) %in% c(samples_AM))
                      # filter(rownames(.) %in% c(samples_AM, samples_AC))
df_pcoa <- dist_matrix_betapart$beta.sor
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa_all, "sorensen")
unique(df_pcoa_new$Country)
my_shape_list <- c("Malaysia" = 20,
                   "India" = 15,
                   "Japan" = 17,
                   "Switzerland" = 18
)
my_size_list <- c("Malaysia" = 8,
                   "India" = 6,
                   "Japan" = 6,
                   "Switzerland" = 8
)
unique(df_pcoa_new$Colony_ID)
my_colony_col_list3 <- c("M1" = "#e31a1c",
                        "M2" = "#cab2d6",
                        "M3" = "#6a3d9a",
                        "M4" = "#41b6c4",
                        "M5" = "#081d58",
                        "M6" = "#225ea8",
                        "M7" = "#3288bd",
                        "AmAi" = "#b2df8a",
                        "AmIu" = "#33a02c",
                        "DrY1" = "#fb9a99",
                        "DrY2" = "#fdbf6f",
                        "GrY2" = "#ff7f00",
                        "C1" = "#cab2d6",
                        "C2" = "#6a3d9a",
                        "C3" = "#54278f",
                        "C4" = "#081d58",
                        "C5" = "#225ea8",
                        "C6" = "#b2df8a",
                        "C7" = "#33a02c",
                        "C8" = "#00441b",
                        "C9" = "#7fcdbb",
                        "AcKn" = "#fdbf6f",
                        "AcCh" = "#ff7f00"

)
metadata <- df_meta_complete %>% select(ID, Colony_ID, Country, Species)
matrix <- as.matrix(dist_matrix_betapart$beta.sor)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new, aes(x = ev1,
                        y = ev2,
                        shape = Country,
                        size = Country,
                        color = Colony_ID))+
                geom_point(stat="identity") +
                stat_ellipse(size = 1) +
                  labs(x=axis_x_title, y = axis_y_title, color = "Colony", shape = "Country") +
                  facet_wrap(~Species, scales = "free") +
                    make_theme(setFill = F, setCol = F, guide_nrow = 3, leg_size = 10 ) +
                      scale_color_manual(values = my_colony_col_list3) +
                      scale_shape_manual(values = my_shape_list) +
                      scale_size_manual(values = my_size_list)
      ggsave("results/figures/09-PCoA_sorensen_microbiome_w_location_colony_AM.pdf", width = 10, height = 15)

# test colony effect in mellifera
adonis_res <- adonis2(dist_matrix_betapart$beta.sor ~ Colony_ID ,data =  df_pcoa_new, permutations = 999, by = "margin")
adonis_OmegaSq(adonis_res)

df_magOTUs_vegan_pa_all <- abundance_df_matrix_pa_all %>%
                      filter(rowSums(.) > 0) %>%
                      filter(rownames(.) %in% c(samples_AC))
                      # filter(rownames(.) %in% c(samples_AM, samples_AC))
# df_pcoa <- dist_matrix_betapart$beta.sor
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa_all, "sorensen")
metadata <- df_meta_complete %>% select(ID, Colony_ID, Country, Species)
matrix <- as.matrix(dist_matrix_betapart$beta.sim)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new, aes(x = ev1,
                        y = ev2,
                        shape = Country,
                        size = Country,
                        color = Colony_ID))+
                geom_point(stat="identity") +
                  labs(x=axis_x_title, y = axis_y_title, color = "Colony", shape = "Country") +
                  facet_wrap(~Species, scales = "free") +
                    make_theme(setFill = F, setCol = F, guide_nrow = 3, leg_size = 10 ) +
                      scale_color_manual(values = my_colony_col_list3) +
                      scale_shape_manual(values = my_shape_list) +
                      scale_size_manual(values = my_size_list)
      ggsave("results/figures/09-PCoA_sorensen_microbiome_w_location_colony_AC.pdf", width = 10, height = 15)
adonis_res <- adonis2(dist_matrix_betapart$beta.sor ~ Colony_ID ,data =  df_pcoa_new, permutations = 999, by = "margin")
adonis_OmegaSq(adonis_res)

df_magOTUs_vegan_pa_all <- abundance_df_matrix_pa_all %>%
                      filter(rowSums(.) > 0) 
# df_pcoa <- dist_matrix_betapart$beta.sor
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa_all, "sorensen")
metadata <- df_meta_complete %>% select(ID, Colony_ID, Country, Species)
matrix <- as.matrix(dist_matrix_betapart$beta.sor)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new, aes(x = ev1,
                        y = ev2,
                        shape = Country,
                        size = Country,
                        color = Colony_ID))+
                geom_point(stat="identity") +
                  labs(x=axis_x_title, y = axis_y_title, color = "Colony", shape = "Country") +
                  # facet_wrap(~Species, scales = "free") +
                    make_theme(setFill = F, setCol = F, guide_nrow = 4, leg_size = 10 ) +
                      scale_color_manual(values = my_colony_col_list3) +
                      scale_shape_manual(values = my_shape_list) +
                      scale_size_manual(values = my_size_list) +
                      new_scale_color() +
                      stat_ellipse(aes(group = Species, color = Species), size = 1) +
                      scale_color_manual(values = host_order_color_dark)
                      ggsave("results/figures/09-PCoA_sorensen_microbiome_w_location_colony_ALL.pdf", width = 10, height = 15)
                      


# beta.sim dist object, dissimilarity matrix accounting for spatial turnover (replacement),
# measured as Simpson pair-wise dissimilarity
# beta.sne dist object, dissimilarity matrix accounting for nestedness-resultant dissimilarity, measured as the nestedness-fraction of Sorensen pair-wise dissimilarity
# beta.sor dist object, dissimilarity matrix accounting for total dissimilarity, measured as
# Sorensen pair-wise dissimilarity (a monotonic transformation of beta diversity)
# For index.family="jaccard" the three matrices are:
# beta.jtu dist dissimilarity matrix accounting for spatial turnover, measured as the turnoverfraction of Jaccard pair-wise dissimilarity
# beta.jne dist object, dissimilarity matrix accounting for nestedness-resultant dissimilarity, measured as the nestedness-fraction of Jaccard pair-wise dissimilarity
# beta.jac dist object, dissimilarity matrix accounting for beta diversity, measured as Jaccard pair-wise dissimilarity (a monotonic transformation of beta diversity)

# Distances from
# https://link.springer.com/article/10.1007/s11756-022-01123-6/tables/4
# 		7	5	3	1	2
# mellifera	7		0.112	0.128	0.132	0.146
# cerana	5	0.112		0.133	0.13	0.119
# dorsata	3	0.128	0.133		0.095	0.108
# florea	1	0.132	0.13	0.095		0.047
# andreniformis	2	0.146	0.119	0.108	0.047	

get_host_divergence <- function(sample_1, sample_2) {
  host_1 <- get_host_from_sample_name(sample_1)
  host_1 <- strsplit(host_1, " ")[[1]][[2]]
  host_2 <- get_host_from_sample_name(sample_2)
  host_2 <- strsplit(host_2, " ")[[1]][[2]]
  mellifera <- c(0, 0.112, 0.128, 0.132, 0.146)
  cerana <- c(0.112, 0, 0.133, 0.13, 0.119)
  dorsata <- c(0.128, 0.133, 0, 0.095, 0.108)
  florea <- c(0.132, 0.13, 0.095, 0, 0.047)
  andreniformis <- c(0.146, 0.119, 0.108, 0.047, 0)
  
  # from literature https://static-content.springer.com/esm/art%3A10.1038%2Fs41598-023-35937-4/MediaObjects/41598_2023_35937_MOESM6_ESM.jpg
        # lengths measured using https://eleif.net/photomeasure#howto for nodes that are not marked
        # 0.127/0.0115 = 11.04 Myr for flo - mel
        # check
        # if set([a, b]) == set(['Apis mellifera', 'Apis cerana']):
        #     return 7.15
        # if set([a, b]) == set(['Apis mellifera', 'Apis dorsata']):
        #     return 9.92
        # if set([a, b]) == set(['Apis mellifera', 'Apis florea']):
        #     return 11.04
        # if set([a, b]) == set(['Apis mellifera', 'Apis andreniformis']):
        #     return 11.04
        # if set([a, b]) == set(['Apis dorsata', 'Apis cerana']):
        #     return 9.92
        # if set([a, b]) == set(['Apis dorsata', 'Apis florea']):
        #     return 11.04
        # if set([a, b]) == set(['Apis dorsata', 'Apis andreniformis']):
        #     return 11.04
        # if set([a, b]) == set(['Apis cerana', 'Apis florea']):
        #     return 11.04
        # if set([a, b]) == set(['Apis cerana', 'Apis andreniformis']):
        #     return 11.04
        # if set([a, b]) == set(['Apis florea', 'Apis andreniformis']):
        #     return 6.42
  # mellifera <- c(0, 7.15, 9.92, 11.04, 11.04)
  # cerana <- c(7.15, 0, 9.92, 11.04, 11.04)
  # dorsata <- c(9.92, 9.92, 0, 11.04, 11.04)
  # florea <- c(11.04, 4, 11.04, 0, 6.42)
  # andreniformis <- c(11.04, 11.04, 11.04, 6.42, 0)

  # mellifera <- c(0, 1, 2, 4, 4)
  # cerana <- c(1, 0, 2, 4, 4)
  # dorsata <- c(2, 2, 0, 3, 3)
  # florea <- c(4, 4, 3, 0, 1)
  # andreniformis <- c(4, 4, 3, 1, 0)
  host_divergence_mat <- rbind(mellifera, cerana, dorsata, florea, andreniformis)
  colnames(host_divergence_mat) <- c("mellifera", "cerana", "dorsata", "florea", "andreniformis")
  index_list <- list("mellifera" = 1, "cerana" = 2, "dorsata" = 3, "florea" = 4, "andreniformis" = 5)
  host_divergence <- host_divergence_mat[as.numeric(index_list[host_1]), as.numeric(index_list[host_2])]
  return(host_divergence)
}
sample_host_div <- data.frame()
for (sample in samples_IN_MY) {
  sample_host_div <- rbind(sample_host_div, rep(0, length((samples_IN_MY))))
}
colnames(sample_host_div) <- samples_IN_MY
rownames(sample_host_div) <- samples_IN_MY
for (sample_1 in samples_IN_MY) {
  for (sample_2 in samples_IN_MY) {
    ind_1 <- as.numeric(which(samples_IN_MY == sample_1))
    ind_2 <- as.numeric(which(samples_IN_MY == sample_2))
    sample_host_div[ind_1, ind_2] <- get_host_divergence(sample_1, sample_2)
  }
}
# get_host_divergence("M1-1", "D1-1")
# get_host_divergence("D3-4", "F1-1")
# get_host_divergence("F1-1", "D3-4")
# samples_subset_am <- df_meta_complete %>% filter(Species == "Apis mellifera") %>% pull(ID)
# samples_subset_ac <- df_meta_complete %>% filter(Species == "Apis cerana") %>% pull(ID)
# samples_subset_ad <- df_meta_complete %>% filter(Species == "Apis dorsata") %>% pull(ID)
# samples_subset_af <- df_meta_complete %>% filter(Species == "Apis florea") %>% pull(ID)
# samples_subset_aa <- df_meta_complete %>% filter(Species == "Apis andreniformis") %>% pull(ID)

df_magOTUs_vegan_IN_MY <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0)
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_IN_MY, "sorensen")
sample_bac_div <- dist_matrix_betapart$beta.sim

heatmap(sample_bac_div %>% as.matrix)
heatmap(sample_host_div %>% as.matrix)

anno_side_b <- rowAnnotation(Host = sample_bac_div %>% as.matrix %>% rownames %>% as.data.frame() %>%
                                mutate(host = Vectorize(get_host_from_sample_name)(.)) %>% pull(host),
                               col = list(Host = host_order_color_dark),
                               border = TRUE
)
anno_side_h <- rowAnnotation(Host = sample_host_div %>% as.matrix %>% rownames %>% as.data.frame() %>%
                                mutate(host = Vectorize(get_host_from_sample_name)(.)) %>% pull(host),
                               col = list(Host = host_order_color_dark),
                               border = TRUE
)
anno_top_b <- HeatmapAnnotation(Host = sample_bac_div %>% as.matrix %>% colnames %>% as.data.frame() %>%
                                mutate(host = Vectorize(get_host_from_sample_name)(.)) %>% pull(host),
                               col = list(Host = host_order_color_dark),
                               border = TRUE
)
anno_top_h <- HeatmapAnnotation(Host = sample_host_div %>% as.matrix %>% colnames %>% as.data.frame() %>%
                                mutate(host = Vectorize(get_host_from_sample_name)(.)) %>% pull(host),
                               col = list(Host = host_order_color_dark),
                               border = TRUE
)
pdf("results/figures/09-Microbe_divergence_dend.pdf", width = 10, height = 15)
Heatmap(sample_bac_div %>% as.matrix,
        rect_gp = gpar(type = "none"),
        top_annotation = anno_top_b,
        right_annotation = anno_side_b
      )
dev.off()
pdf("results/figures/09-Host_divergence_dend.pdf", width = 10, height = 15)
Heatmap(sample_host_div %>% as.matrix,
        rect_gp = gpar(type = "none"),
        top_annotation = anno_top_h,
        right_annotation = anno_side_h
      )
dev.off()

sample_host_div <- sample_host_div %>% 
  filter(row.names(.) %in% as.vector(as.matrix(sample_bac_div) %>% row.names))

sample_host_div_df <- sample_host_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                        pivot_longer(!Sample, names_to = "Compared", values_to = "dist_h")
sample_bac_div_df <- sample_bac_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                          pivot_longer(!Sample, names_to = "Compared", values_to = "dist_b") %>%
                          mutate(Host_s = Vectorize(get_host_from_sample_name)(Sample)) %>%
                          mutate(Host_c = Vectorize(get_host_from_sample_name)(Compared)) %>%
                          mutate(Type_b = ifelse(Host_s == Host_c, "Within", "Between")) %>%
                          mutate(Hosts_compared = paste0(Host_s, "_", Host_c)) %>%
                          filter(Host_s != Host_c) # added only here at the figure creation stage if using other distances make sure to do this too there
dissimilarities_df_median <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, median_b = median(dist_b)) %>%
                        unique
dissimilarities_df_mean <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, mean_b = median(dist_b)) %>%
                        unique
dissimilarities_df <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared)

mantel(sample_host_div %>% as.dist, sample_bac_div, method="pearson", permutations=1000)

cor_res <- cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "spearman")
# dissimilarities_df_filt <- dissimilarities_df %>%
#                             filter(Sample != Compared)
# cor_res <- cor.test(dissimilarities_df_filt$dist_h, dissimilarities_df_filt$dist_b, method = "spearman")
# ggplot() +
#   geom_violin(data=dissimilarities_df,
#              aes(x = dist_h,
#                  y = dist_b,
#                  group = Hosts_compared,
#                  fill = Host_c,
#                 #  color = Host_s,
#                 #  shape = Type_b
#              ), width=0.015, 
#              position = "identity",
#              alpha = 0.5
#             ) +
#   geom_violin(data=dissimilarities_df,
#              aes(x = dist_h,
#                  y = dist_b,
#                  group = Hosts_compared,
#                  fill = Host_s,
#                 #  color = Host_s,
#                 #  shape = Type_b
#              ), width=0.015, 
#              position = "identity",
#              alpha = 0.5
#             ) +
#     labs(x = "Host divergence (# nodes between)", y = "Microbiome dissimilarity (jaccard)") +
#     scale_fill_manual(values = host_order_color) +
#     scale_color_manual(values = host_order_color_dark) +
#     make_theme(setCol = F, setFill = F, guide_nrow = 1)
#     ggsave("results/figures/09-Microbe_host_dissimilarities_violin_sorenson.pdf")

ggplot() +
  geom_jitter(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_s,
                 color = Host_c,
                #  shape = Type_b
             ), width=0, size = 4, shape = 21, stroke = 2
            ) +
  # geom_jitter(data=dissimilarities_df_mean,
  #            aes(x = dist_h,
  #                y = mean_b
  #            ), size = 3, width = 0
  #           ) +
  geom_jitter(data=dissimilarities_df_median,
             aes(x = dist_h,
                 y = median_b
             ), size = 5, width = 0, color = "black", shape = 23, fill = "#ffffff", stroke = 2
            ) +
  # geom_smooth(data=dissimilarities_df_median,
  #            aes(x = dist_h,
  #                y = median_b
  #            ), method = "lm", se = T, color = "black"
  #           ) +
    labs(x = "Host divergence (based on 16S and 12S)", y = "Microbiome dissimilarity (sorensen - turnover)", color = "Host species") +
    scale_fill_manual(values = host_order_color) +
    scale_color_manual(values = host_order_color) +
    make_theme(setCol = F, setFill = F, guide_nrow = 2,
               leg_size = 12,
               axis_x_title = 20, axis_y_title = 20
    ) +
  annotate("text", x = 0.001, y = 0.95, label = paste0("Spearman's rho = ", round(cor_res$estimate, 2), "\n p-value = ", cor_res$p.value), size = 5, color = "black", hjust = 0, vjust = 1)
  ggsave("results/figures/09-figures/09-Microbe_host_dissimilarities_sorenson_turnover.pdf", width = 15, height = 15)
  # ggsave("results/figures/09-Microbe_host_dissimilarities_sorenson_turnover_with_same.pdf")

summary(lm(dissimilarities_df_median$median_b ~ dissimilarities_df_median$dist_h))
mantel(as.dist(sample_host_div), as.dist(sample_bac_div), method = "pearson", permutations = 9999)

# Permute 5 x 5 for testing:
# select 5 random samples from samples
signifs <- c()
r_stat <- c()
res_df <- data.frame()
perms <- 1000
for (n in 1:perms) {
  select_am <- sample(intersect(samples_AM, samples_IN_MY), 1)
  select_ac <- sample(intersect(samples_AC, samples_IN_MY), 1)
  select_ad <- sample(intersect(samples_AD, samples_IN_MY), 1)
  select_af <- sample(intersect(samples_AF, samples_IN_MY), 1)
  select_aa <- sample(intersect(samples_AA, samples_IN_MY), 1)
  selected_samples <- c(select_am, select_ac, select_ad, select_af, select_aa)
  mat_bac_div <- sample_bac_div %>% as.matrix
  select_bac_div <- mat_bac_div[selected_samples, selected_samples] %>% as.dist
  select_host_div <- sample_host_div[selected_samples, selected_samples] %>% as.dist
  # heatmap(select_bac_div %>% as.matrix)
  # heatmap(select_host_div %>% as.matrix)
  res_man <- mantel(select_host_div, select_bac_div, method = "pearson")
  signifs <- c(signifs, res_man$signif)
  r_stat <- c(r_stat, res_man$statistic)
  samples_str <- paste0(selected_samples, collapse = "|")
  res_df <- rbind(res_df, cbind("samples" = samples_str, "stat" = res_man$statistic, "sig" = res_man$signif))
}
adj_signifs <- p.adjust(signifs, method = "fdr", n = perms)
median_p <- median(signifs)
median_p_adj <- median(adj_signifs)
mean_p_adj <- mean(adj_signifs)
sd_p_adj <- sd(adj_signifs)
min_p_adj <- min(adj_signifs)
max_p_adj <- max(adj_signifs)
median_r <- median(r_stat)
mean_r <- mean(r_stat)
sd_r <- sd(r_stat)
write.csv(res_df, "results/figures/09-figures/09-Mantel_permutations_phylosymbiosis.csv")
hist(signifs, 100)
# how many are significant?
length(adj_signifs[adj_signifs < 0.05])
hist(adj_signifs, 50)

split_by_pattern <- function(my_name, pattern, index) {
  return(strsplit(my_name, pattern)[[1]][[index]])
}
bac_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(dist_h, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = median_b) %>%
                    column_to_rownames("Compared1")
host_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(median_b, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = dist_h) %>%
                    column_to_rownames("Compared1")

bac_median_mat[1,]
host_median_mat[1,]

mantel(as.dist(bac_median_mat), as.dist(host_median_mat), method = "pearson", permutations = 9999)


# sample_bac_div <- dist_matrix_atch
sample_bac_div <- dist_matrix_atch_rel
sample_host_div <- data.frame()
for (sample in samples) {
  sample_host_div <- rbind(sample_host_div, rep(0, length((samples))))
}
colnames(sample_host_div) <- samples
rownames(sample_host_div) <- samples
for (sample_1 in samples) {
  for (sample_2 in samples) {
    ind_1 <- as.numeric(which(samples == sample_1))
    ind_2 <- as.numeric(which(samples == sample_2))
    sample_host_div[ind_1, ind_2] <- get_host_divergence(sample_1, sample_2)
  }
}

# samples_subset_am <- df_meta_complete %>% filter(Species == "Apis mellifera") %>% pull(ID)
# samples_subset_ac <- df_meta_complete %>% filter(Species == "Apis cerana") %>% pull(ID)
# samples_subset_ad <- df_meta_complete %>% filter(Species == "Apis dorsata") %>% pull(ID)
# samples_subset_af <- df_meta_complete %>% filter(Species == "Apis florea") %>% pull(ID)
# samples_subset_aa <- df_meta_complete %>% filter(Species == "Apis andreniformis") %>% pull(ID)

heatmap(sample_bac_div %>% as.matrix)

sample_host_div <- sample_host_div %>% 
  filter(row.names(.) %in% as.vector(as.matrix(sample_bac_div) %>% row.names))

sample_host_div_df <- sample_host_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                        pivot_longer(!Sample, names_to = "Compared", values_to = "dist_h")
sample_bac_div_df <- sample_bac_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                          pivot_longer(!Sample, names_to = "Compared", values_to = "dist_b") %>%
                          mutate(Host_s = Vectorize(get_host_from_sample_name)(Sample)) %>%
                          mutate(Host_c = Vectorize(get_host_from_sample_name)(Compared)) %>%
                          mutate(Type_b = ifelse(Host_s == Host_c, "Within", "Between")) %>%
                          mutate(Hosts_compared = paste0(Host_s, "_", Host_c))
dissimilarities_df_median <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, median_b = median(dist_b)) %>%
                        unique
dissimilarities_df_mean <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, mean_b = median(dist_b)) %>%
                        unique
dissimilarities_df <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared)
                        # filter(Sample %in% c("M1.2", "C1.2", "D1.2", "F1.2", "A1.2"))

mantel(sample_host_div, sample_bac_div, method="pearson", permutations=999)

ggplot() +
  geom_violin(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_c,
                #  color = Host_s,
                #  shape = Type_b
             ), width=0.015, 
             position = "identity",
             alpha = 0.5
            ) +
  geom_smooth(data = dissimilarities_df, aes(x = dist_h,
                 y = dist_b), color = "#000000", method = "lm", se = FALSE) +
  geom_violin(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_s,
                #  color = Host_s,
                #  shape = Type_b
             ), width=0.015, 
             position = "identity",
             alpha = 0.5
            ) +
    labs(x = "Host divergence (# nodes between)", y = "Microbiome dissimilarity (jaccard)") +
    scale_fill_manual(values = host_order_color) +
    scale_color_manual(values = host_order_color_dark) +
    make_theme(setCol = F, setFill = F, guide_nrow = 1)
    ggsave("results/figures/09-Microbe_host_dissimilarities_violin_atchinson.pdf")

ggplot() +
  geom_jitter(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_s,
                 color = Host_c,
                #  shape = Type_b
             ), width=0.001, size = 4, shape = 21, stroke = 2
            ) +
  # geom_jitter(data=dissimilarities_df_mean,
  #            aes(x = dist_h,
  #                y = mean_b
  #            ), size = 3, width = 0
  #           ) +
  geom_jitter(data=dissimilarities_df_median,
             aes(x = dist_h,
                 y = median_b
             ), size = 3, width = 0
            ) +
    labs(x = "Host divergence (based on 16S and 12S)", y = "Microbiome dissimilarity (jaccard)", color = "Host species") +
    scale_fill_manual(values = host_order_color_dark) +
    scale_color_manual(values = host_order_color_dark) +
    make_theme(setCol = F, setFill = F, guide_nrow = 2,
               leg_size = 12,
               axis_x_title = 15, axis_y_title = 15
    )
  ggsave("results/figures/09-Microbe_host_dissimilarities_atchinson.pdf")
cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")

split_by_pattern <- function(my_name, pattern, index) {
  return(strsplit(my_name, pattern)[[1]][[index]])
}
bac_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(dist_h, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = median_b) %>%
                    column_to_rownames("Compared1")
host_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(median_b, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = dist_h) %>%
                    column_to_rownames("Compared1")

bac_median_mat[1,]
host_median_mat[1,]

mantel(as.dist(bac_median_mat), as.dist(host_median_mat), method = "pearson", permutations = 9999)
cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")

# edit tree to look prettier later
# plot_tree(ps, nodelabf=nodeplotblank, color="Host", label.tips = "Name", ladderize="left", plot.margin=0.3) + 
#   scale_color_manual(values = host_order_color)
#   ggsave("results/figures/09-Plot_bac120_tree_with_host.pdf")


# Is phylogenetic distance correlated with presence/absence (rel ab) correlation?



```

# number of reads used for profiling and rarefaction curves

```{r}
# abundance_df_matrix_copies_int <- abundance_df_matrix_copies %>% as.matrix %>% as.data.frame %>% mutate_all(as.integer)
# S <- specnumber(abundance_df_matrix_copies_int)
# raremax <- min(rowSums(abundance_df_matrix_copies_int))
# Srare <- rarefy(abundance_df_matrix_copies_int, raremax)
# pdf("results/figures/09-rarefaction_curve_test.pdf")
# rarecurve(abundance_df_matrix_copies_int, step = 20, 
#           sample = raremax, 
#           col = "blue", 
#           cex = 0.6,
#           main = "rarecurve()")
# dev.off()

# reads used by instrain:
df_mapping_info <- data.frame()
for (sample in samples_IN_MY) {
  mapping_info_path <- paste0("results/10_instrain/02_instrain_profile/", sample, "/output/", sample, "_mapping_info.tsv")
  if (file.exists(mapping_info_path)) {
    print(paste0("Reading ", sample))
    df_mapping <- read.csv(mapping_info_path, sep = "\t", skip = 1, nrows=1)
    df_mapping_info <- bind_rows(df_mapping_info, cbind(df_mapping, "sample"=sample))
  }
}

ggplot(df_mapping_info %>% mutate(Host = Vectorize(get_host_from_sample_name)(sample)), 
      aes(x = filtered_pairs,
          y = sample,
          fill = Host)) +
  geom_bar(stat = "identity") +
  labs(x = "Number of reads", y = "Sample") +
  scale_fill_manual(values = host_order_color) +
  scale_x_continuous(labels = scales::comma) +
  make_theme(setCol = F, setFill = F, guide_nrow = 1, y_size = 7)
  ggsave("results/figures/09-number_of_reads_filtered_instrain.pdf")

# correlation between number of reads and number of species detected?


```

# Bar plots by genus

```{r}
# plot abundance by genus

# colnames(vis_magOTUs_df)
# handmade_species_names <- read.csv("config/Species_MAG_Cluster-names.txt", header = T, sep = "\t")
# df_plot <- abundance_df %>% filter(!is.na(magOTU)) %>%
#                             filter(breadth > breadth_cutoff) %>%
#                             mutate(detected = ifelse(coverage > 1 & breadth > breadth_cutoff, 1, 0)) %>%
#                             select(sample, Host, magOTU, detected, Genus, coverage, copies_cov, rel_cov) %>%
#                             left_join(prevs_df, by = c("magOTU" = "X", "Host" = "Host")) %>%
#                             select(magOTU, coverage, copies_cov, rel_cov, prevalence_in_host, Genus, Host, sample) %>%
#                             left_join(handmade_species_names %>% select(MAG_species_name_final, cluster_name_instrain), by = c("magOTU" = "cluster_name_instrain"))
rel_p <- ggplot(df_plot) +
  geom_bar(aes(x = sample,
               y = rel_cov,
               fill = Genus), stat = "identity") +
  # facet_grid(Host ~ ., scale = "free_x") +
  facet_grid(~Host, scale = "free_x") +
  labs(y = "Sample", x = "Relative abundance") +
  scale_fill_manual(values=genusColors) +
  make_theme(theme_name=theme_few(), leg_pos="none",
             # guide_nrow = 6,
             setFill = F, setCol = F,
             x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
             y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
ggsave(rel_p, "results/figures/09-relative_abundance_sep.pdf")
pdf("results/figures/09-relative_abundance_sep_legend.pdf")
plot(get_only_legend(rel_p + make_theme(setCol = F, setFill = F, guide_nrow = 10, leg_size = 8, leg_pos = "right")))
dev.off()


system("mkdir -p results/figures/Species_Relative_abundance")

# p_list <- list()
for (genus in unique(df_plot$Genus)) {
  if (is.na(genus)) {
    next
  }
  p <- ggplot(df_plot %>% filter(Genus == genus) %>% group_by(sample) %>% mutate(rel_cov = rel_cov/sum(rel_cov))) +
    geom_bar(aes(x = sample,
                 y = rel_cov,
                 fill = MAG_species_name_final), color = "black", linewidth = 0.1, stat = "identity") +
    # facet_grid(factor(Host, host_order) ~ ., scale = "free_x") +
    facet_grid(~factor(Host, host_order), scale = "free_x") +
    ggtitle(genus) +
    labs(y = "Sample", x = "Relative abundance", fill = "Species") +
    scale_fill_manual(values=head(colorRampPalette(c(genusColors[[genus]], "#FFFFFF"))(2+length(unique(df_plot %>% filter(Genus == genus) %>% pull(MAG_species_name_final)))), -2)) +
    # scale_fill_manual(values=genusColors) +
    make_theme(theme_name=theme_few(), leg_pos="right",
               guide_nrow = 10,
               setFill = F, setCol = F,
               x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
               y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
  ggsave(paste0("results/figures/Species_Relative_abundance/",genus, "_spec_rel_abund.pdf"), plot = p, width = 25, height = 15)
  # p_list[[genus]] <- p
}

# pdf("results/figures/09-relative_abundance_by_genus.pdf", width = 15, height = 10)
# marrangeGrob(p_list, nrow = 2, ncol = 1)
# dev.off()



# p_list <- list()
for (genus in unique(df_plot$Genus)) {
  if (is.na(genus)) {
    next
  }
  p <- ggplot(df_plot %>% filter(Genus == genus) %>% group_by(sample) %>% mutate(rel_cov = rel_cov/sum(rel_cov))) +
    geom_bar(aes(x = sample,
                 y = rel_cov,
                 fill = MAG_species_name_final), color = "black", linewidth = 0.1, stat = "identity") +
    # facet_grid(factor(Host, host_order) ~ ., scale = "free_x") +
    facet_grid(~factor(Host, host_order), scale = "free_x") +
    ggtitle(genus) +
    labs(y = "Sample", x = "Relative abundance", fill = "Species") +
    scale_fill_manual(values=make_col_list(df_plot %>% filter(Genus == genus) %>% pull(MAG_species_name_final))) +
    # scale_fill_manual(values=genusColors) +
    make_theme(theme_name=theme_few(), leg_pos="right",
               guide_nrow = 10,
              #  guide_nrow = length(unique(df_plot %>% filter(Genus == genus) %>% pull(magOTU))),
               setFill = F, setCol = F,
               x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
               y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
  ggsave(paste0("results/figures/Species_Relative_abundance/",genus, "_spec_rel_abund_dist_colors.pdf"), plot = p, width = 20, height = 15)
  # p_list[[genus]] <- p
}

# pdf("results/figures/09-relative_abundance_by_genus_dist_colors.pdf", width = 15, height = 10)
# marrangeGrob(p_list, nrow = 2, ncol = 1)
# dev.off()

# p_list <- list()
for (genus in unique(df_plot$Genus)) {
  if (is.na(genus)) {
    next
  }
  p <- ggplot(df_plot %>% filter(Genus == genus) %>% group_by(sample) %>% mutate(rel_cov = rel_cov/sum(rel_cov))) +
    geom_bar(aes(x = sample,
                 y = coverage,
                 fill = MAG_species_name_final), color = "black", linewidth = 0.1, stat = "identity") +
    # facet_grid(factor(Host, host_order) ~ ., scale = "free_x") +
    facet_grid(~factor(Host, host_order), scale = "free_x") +
    ggtitle(genus) +
    labs(y = "Sample", x = "Relative abundance", fill = "Species") +
    scale_fill_manual(values=make_col_list(df_plot$MAG_species_name_final)) +
    scale_y_continuous(trans = "log") +
    # scale_fill_manual(values=genusColors) +
    make_theme(theme_name=theme_few(), leg_pos="right",
               guide_nrow = 10,
              #  guide_nrow = length(unique(df_plot %>% filter(Genus == genus) %>% pull(magOTU))),
               setFill = F, setCol = F,
               x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
               y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
  # p_list[[genus]] <- p
  ggsave(paste0("results/figures/Species_Relative_abundance/",genus, "_spec_coverage_dist_colors.pdf"), plot = p, width = 20, height = 15)
}

# pdf("results/figures/09-abundance_coverage_by_genus_dist_colors.pdf", width = 15, height = 10)
# marrangeGrob(p_list, nrow = 2, ncol = 1)
# dev.off()

for (genus in unique(df_plot$Genus)) {
  if (is.na(genus)) {
    next
  }
  p <- ggplot(df_plot %>% filter(Genus == genus)) +
    geom_bar(aes(x = sample,
                 y = rel_cov,
                 fill = magOTU), color = "white", stat = "identity") +
    # facet_grid(factor(Host, host_order) ~ ., scale = "free_x") +
    facet_grid(~factor(Host, host_order), scale = "free_x") +
    ggtitle(genus) +
    labs(y = "Sample", x = "Relative abundance") +
    scale_fill_manual(values=head(colorRampPalette(c(genusColors[[genus]], "#FFFFFF"))(2+length(unique(df_plot %>% filter(Genus == genus) %>% pull(magOTU)))), -2)) +
    # scale_fill_manual(values=genusColors) +
    make_theme(theme_name=theme_few(), leg_pos="none",
               guide_nrow = 10,
              #  guide_nrow = length(unique(df_plot %>% filter(Genus == genus) %>% pull(magOTU))),
               setFill = F, setCol = F,
               x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
               y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
  ggsave(paste0("results/figures/Species_Relative_abundance/",genus, "_spec_rel_abund_dist_colors_no_legend.pdf"), plot = p, width = 20, height = 15)
  # p_list[[genus]] <- p
}

# pdf("results/figures/09-relative_abundance_sep_by_genus.pdf", width = 15, height = 10)
# marrangeGrob(p_list, nrow = 2, ncol = 1)
# dev.off()


ggplot(df_plot %>% select(magOTU, prevalence_in_host, sample, Host, Genus, rel_cov) %>%
          group_by(magOTU, Host) %>%
          mutate(rel_cov_mean = mean(rel_cov)) %>%
          mutate(rel_cov_sd = sd(rel_cov)) %>%
          ungroup() %>%
          select(magOTU, prevalence_in_host, rel_cov_mean, rel_cov_sd, Genus, Host) %>% unique
) +
  geom_point(aes(x = prevalence_in_host,
               y = rel_cov_mean), size = 5, color = "black") +
  geom_point(aes(x = prevalence_in_host,
               y = rel_cov_mean,
               color = Genus), size = 4) +
  facet_wrap(~Host, scales = "free", ncol = 2) +
  geom_errorbar(aes(x = prevalence_in_host,
                    ymin = rel_cov_mean - rel_cov_sd,
                    ymax = rel_cov_mean + rel_cov_sd,
                    color = Genus)) +
  geom_smooth(aes(x = prevalence_in_host,
                  y = rel_cov_mean), method = "glm") +
  stat_cor(aes(x = prevalence_in_host,
                  y = rel_cov_mean), method = "spearman") +
  labs(x = "Prevalence in host", y = "Relative abundance (mean in Host)") +
  scale_color_manual(values=genusColors) +
  make_theme(theme_name=theme_minimal(), leg_pos="none",
             # guide_nrow = 6,
             setFill = F, setCol = F,
             x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
             y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
ggsave("results/figures/09-prevalence_vs_rel_abundance.pdf")
```



```{r other_plots}
ggplot(instrain_genomes_info_df %>%
        group_by(genome) %>%
          mutate(breadth_pass = ifelse(breadth >= 0.7, "Y", "N")) %>%
          mutate(perc_var = SNV_count/length*breadth*100) %>%
          filter(breadth_pass %in% c("Y")),
                             aes(x = magOTU,
                                 y = perc_var,
                                 color = factor(Host, host_order),
                                 shape = breadth_pass
                              )
                            ) +
      geom_jitter() +
      labs(x = "magOTU", y = "% variant sites", color = "Host species") +
        facet_wrap(~factor(Genus, shared_genera), scales = "free_x") +
          make_theme(theme_name=theme_few(), leg_pos="bottom",
                     guide_nrow = 3,
                     setFill = F, setCol = F,
                     x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                     y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
            scale_color_manual(values=host_order_color)
            ggsave("results/figures/15b-percentage_variable_sites.pdf")

```

# "timescale" of evolution

```{r}
get_stat_label <- function(df_in) {
  # df_in = df_plot
  # lm_res <- lm(mean_dist ~ Host_time, data = df_in)
  lm_res <- lm(Bact_dist ~ Host_time, data = df_in)
  slope_coef = coef(lm_res)[2]
  r2 = summary(lm_res)$adj.r.squared
  p = summary(lm_res)$coefficients[,4][2]
  # x_max <- max(df_in$Host_dist)/2
  x_max <- max(df_in$Host_time)/2
  # y_max <- max(df_in$mean_dist)
  y_max <- max(df_in$Bact_dist)
  return(list("x" = x_max, "y" = y_max, "slope" = slope_coef, "r" = r2, "p" = p))
}
get_list_label <- function(df_in) {
  x_max <- max(df_in$Host_time)/2
  # y_max <- max(df_in$mean_dist)
  y_min <- min(df_in$Bact_dist)
  return(list("x" = x_max, "y" = y_min))
}
df_phy_dist <- read.csv(paste0("results/figures/12-phylo_distances/bact_host_dists.tsv"), sep = "\t") %>%
                  mutate(spec_comb = ifelse(Species1 == Species2, "Same", "Diff")) %>%
                  filter(Host1 != "Other") %>%
                  filter(Host2 != "Other") %>%
                  # replace na in host time with 0
                  mutate(Host_time = ifelse(Host_time == "None", 0, Host_time)) %>%
                  mutate(Host_time = as.numeric(Host_time))
df_phy_dist_same_spec <- df_phy_dist %>% filter(Species1 == Species2)
df_phy_dist_same_spec %>% filter(grepl("Bombi", Species1))
p_list <- list()
for (species in unique(df_phy_dist_same_spec$Species1)) {
    df_plot <- df_phy_dist_same_spec %>% filter(Species1 == species)
    if (length(unique(df_plot$Host1)) == 1) {
      next
    }
    p <- ggplot(df_plot) +
      geom_point(aes(x = Host_time,
                    y = Bact_dist,
                    fill = Host2,
                    color = Host1),
                  shape = 21, size = 3, stroke = 3
                  ) +
      # geom_point(aes(x = Host_time,
      #               y = Bact_dist,
      #               color = Host1),
      #             shape = "\u25D2", size = 3, stroke = 3
      #             ) +
      # geom_point(aes(x = Host_time,
      #               y = Bact_dist,
      #               color = Host2),
      #             shape = "\u25D3", size = 3, stroke = 3
      #             ) +
      geom_smooth(aes(x = Host_time,
                    y = Bact_dist), color = "#000000", method = "lm", se = F) +
      annotate("text", x = get_stat_label(df_plot)[["x"]],
                    y = get_stat_label(df_plot)[["y"]],
                    size = 5,
        label = paste0("r2 = ", round(get_stat_label(df_plot)[["r"]], 4), ", p = ", round(get_stat_label(df_plot)[["r"]], 3), ", Slope = ", round(get_stat_label(df_plot)[["slope"]], 5))) +
      labs(x = "Host divergence", y = "Bacteria divergence", title = paste0(species)) +
      scale_color_manual(values = host_order_color_dark) +
      scale_fill_manual(values = host_order_color_dark) +
      scale_x_continuous(labels=unit_format(unit = "Mya", scale = 1)) +
      make_theme(theme_name=theme_few(), leg_pos="none",
                # guide_nrow = 6,
                setFill = F, setCol = F,
                x_angle=45 ,x_vj=1, x_hj=1, x_size=10,
                y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
    p_list[[species]] <- p
  }
pdf(paste0("results/figures/12-phylo_distances/bact_host_dists_same_species.pdf"), width = 15, height = 10)
print(marrangeGrob(p_list, nrow = 2, ncol = 2))
dev.off()


df_phy_dist <- read.csv(paste0("results/figures/12-phylo_distances/bact_host_dists.tsv"), sep = "\t") %>%
                  mutate(spec_comb = ifelse(Species1 == Species2, "Same", "Diff")) %>%
                  filter(Host1 != "Other") %>%
                  filter(Host2 != "Other") %>%
                  # replace na in host time with 0
                  mutate(Host_time = ifelse(Host_time == "None", 0, Host_time)) %>%
                  mutate(Host_time = as.numeric(Host_time))
all_species <- unique(df_phy_dist$Species1)
species1="Bifidobacterium_SGB-6_4"
species2="Bifidobacterium_SGB-6_3"
"Bifidobacterium_sp945269915" %in% all_species
species_l=c("Bifidobacterium_SGB-6_4", "Bifidobacterium_SGB-6_3", "Bifidobacterium_asteroides_I")
df_plot <- df_phy_dist %>% 
                  filter(Species1 %in% species_l) %>%
                  filter(Species2 %in% species_l)
ggplot(df_plot) +
        geom_point(aes(x = Host_time,
                      # y = mean_dist,
                      y = Bact_dist,
                      fill = Host2,
                      color = Host1),
                    shape = 21, size = 3, stroke = 3
                    ) +
        geom_smooth(aes(x = Host_time,
                      # y = mean_dist), color = "#000000", method = "lm", se = F) +
                      y = Bact_dist), color = "#000000", method = "lm", se = F) +
        annotate("text", x = get_stat_label(df_plot)[["x"]],
                      y = get_stat_label(df_plot)[["y"]],
                      size = 10,
          label = paste0("r2 = ", round(get_stat_label(df_plot)[["r"]], 4), ", p = ", round(get_stat_label(df_plot)[["r"]], 3), ", Slope = ", round(get_stat_label(df_plot)[["slope"]], 5))) +
        # annotate("text", x = get_list_label(df_plot)[["x"]],
        #               y = get_list_label(df_plot)[["y"]],
        #               size = 2,
        #   label = paste(unique(df_plot$Species2), collapse = " ")) +
        labs(x = "Host divergence", y = "Bacteria divergence", title = paste(species1, species2)) +
        scale_color_manual(values = host_order_color_dark) +
        scale_fill_manual(values = host_order_color_dark) +
        scale_x_continuous(labels=unit_format(unit = "Mya", scale = 1)) +
        make_theme(theme_name=theme_few(), leg_pos="none",
                  # guide_nrow = 6,
                  setFill = F, setCol = F,
                  axis_x_title = 25, axis_y_title = 25,
                  x_angle=45 ,x_vj=1, x_hj=1, x_size=20,
                  y_angle=0 ,y_vj=0, y_hj=0, y_size=20)
        ggsave(paste0("results/figures/12-phylo_distances/", species1, "_", species2,"_bact_host_dists.pdf"), plot = p, width = 15, height = 20)

# df_phy_dist_same_spec <- df_phy_dist %>% filter(grepl(genus, Species1))
# df_phy_dist_same_spec <- df_phy_dist %>% filter(grepl(genus, Species2))
# df_phy_dist %>%
#                           mutate(indicate1 = ifelse(grepl(genus, Species1), 1, 0)) %>%
#                           mutate(indicate2 = ifelse(grepl(genus, Species2), 1, 0)) %>%
#                             mutate(indicate = indicate1 + indicate2) %>%
#                               filter(indicate == 2) %>%
#                               select(-indicate1, -indicate2, -indicate) %>%
#                               pull(Species1) %>% unique
genus <- "Bifidobacterium"
genus <- "Lactobacillus"
genus <- "Gilliamella"
genus <- "Snodgrassella"
genus <- "Bartonella"
df_plot <- df_phy_dist %>%
                          mutate(indicate1 = ifelse(grepl(genus, Species1), 1, 0)) %>%
                          mutate(indicate2 = ifelse(grepl(genus, Species2), 1, 0)) %>%
                            mutate(indicate = indicate1 + indicate2) %>%
                              filter(indicate == 2) %>%
                              select(-indicate1, -indicate2, -indicate) %>%
                              group_by(Species1, Species2) %>%
                              arrange(Species1, Species2) %>%
                              summarise() %>%
                              write.csv(paste0("results/figures/12-phylo_distances/", genus,"_bact_host_dists.csv"), row.names = F)


# genus <- "Bifidobacterium"
for (genus in genera_clean) {
    # select only species that are host-specific (?)
    # df_plot <- df_phy_dist %>%
    #                       mutate(indicate1 = ifelse(grepl(genus, Species1), 1, 0)) %>%
    #                       mutate(indicate2 = ifelse(grepl(genus, Species2), 1, 0)) %>%
    #                         mutate(indicate = indicate1 + indicate2) %>%
    #                           filter(indicate == 2) %>%
    #                           select(-indicate1, -indicate2, -indicate)
    # if (length(unique(df_plot$Host1)) <= 1) {
    #   next
    # }
    genus <- "Bifidobacterium"
    genus <- "Lactobacillus"
    genus <- "Gilliamella"
    genus <- "Snodgrassella"
    genus <- "Bartonella"
    genus <- "Pectinatus"
    df_plot_top <- df_phy_dist %>%
                          mutate(indicate1 = ifelse(grepl(genus, Species1), 1, 0)) %>%
                          mutate(indicate2 = ifelse(grepl(genus, Species2), 1, 0)) %>%
                            mutate(indicate = indicate1 + indicate2) %>%
                              filter(indicate == 2) %>%
                              select(-indicate1, -indicate2, -indicate) #%>%
                              # group_by(Species1, Species2, Host_time, Host1, Host2) %>%
                              # summarise(mean_dist = mean(Bact_dist))
    if (length(unique(df_plot_top$Species1)) <= 1) {
      next
    }
    p_sublist <- list()
    species1="Lactobacillus_SGB-33_1"
    species2="Lactobacillus_SGB-33_1"
    for (species in unique(df_plot_top$Species1)) {
      df_plot <- df_plot_top %>% 
                  filter(Species1 == species1) %>%
                    filter(Species2 == species2)
      p <- ggplot(df_plot) +
        geom_point(aes(x = Host_time,
                      # y = mean_dist,
                      y = Bact_dist,
                      fill = Host2,
                      color = Host1),
                    shape = 21, size = 3, stroke = 3
                    ) +
        # geom_point(aes(x = Host_time,
        #               y = Bact_dist,
        #              # y = mean_dist,
        #               color = Host1),
        #             shape = "\u25D2", size = 3, stroke = 3
        #             ) +
        # geom_point(aes(x = Host_time,
        #              # y = mean_dist,
        #               y = Bact_dist,
        #               color = Host2),
        #             shape = "\u25D3", size = 3, stroke = 3
        #             ) +
        geom_smooth(aes(x = Host_time,
                      # y = mean_dist), color = "#000000", method = "lm", se = F) +
                      y = Bact_dist), color = "#000000", method = "lm", se = F) +
        annotate("text", x = get_stat_label(df_plot)[["x"]],
                      y = get_stat_label(df_plot)[["y"]],
                      size = 5,
          label = paste0("r2 = ", round(get_stat_label(df_plot)[["r"]], 4), ", p = ", round(get_stat_label(df_plot)[["r"]], 3), ", Slope = ", round(get_stat_label(df_plot)[["slope"]], 5))) +
        annotate("text", x = get_list_label(df_plot)[["x"]],
                      y = get_list_label(df_plot)[["y"]],
                      size = 2,
          label = paste(unique(df_plot$Species2), collapse = " ")) +
        labs(x = "Host divergence", y = "Bacteria divergence", title = paste0(species)) +
        scale_color_manual(values = host_order_color_dark) +
        scale_fill_manual(values = host_order_color_dark) +
        scale_x_continuous(labels=unit_format(unit = "Mya", scale = 1)) +
        make_theme(theme_name=theme_few(), leg_pos="none",
                  # guide_nrow = 6,
                  setFill = F, setCol = F,
                  x_angle=45 ,x_vj=1, x_hj=1, x_size=10,
                  y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
      p_sublist[[species]] <- p
    }
    # pdf(paste0("results/figures/12-phylo_distances/", genus,"_bact_host_dists.pdf"), width = 15, height = 20)
    #   marrangeGrob(p_sublist, nrow = 2, ncol = 2)
    # dev.off()
  }


p_list <- list()
for (genus in genera_clean) {
    # select only species that are host-specific (?)
    # df_plot <- df_phy_dist %>%
    #                       mutate(indicate1 = ifelse(grepl(genus, Species1), 1, 0)) %>%
    #                       mutate(indicate2 = ifelse(grepl(genus, Species2), 1, 0)) %>%
    #                         mutate(indicate = indicate1 + indicate2) %>%
    #                           filter(indicate == 2) %>%
    #                           select(-indicate1, -indicate2, -indicate)
    df_plot <- df_phy_dist %>%
                          mutate(indicate1 = ifelse(grepl(genus, Species1), 1, 0)) %>%
                          mutate(indicate2 = ifelse(grepl(genus, Species2), 1, 0)) %>%
                            mutate(indicate = indicate1 + indicate2) %>%
                              filter(indicate == 2) %>%
                              select(-indicate1, -indicate2, -indicate) #%>%
                              # group_by(Species1, Species2, Host_time, Host1, Host2) %>%
                              # summarise(mean_dist = mean(Bact_dist))
    if (length(unique(df_plot$Host1)) <= 1) {
      next
    }
    p <- ggplot(df_plot) +
      geom_point(aes(x = Host_time,
                    # y = mean_dist,
                    y = Bact_dist,
                    fill = Host2,
                    color = Host1),
                  shape = 21, size = 3, stroke = 3
                  ) +
      # geom_point(aes(x = Host_time,
      #               y = Bact_dist,
      #              # y = mean_dist,
      #               color = Host1),
      #             shape = "\u25D2", size = 3, stroke = 3
      #             ) +
      # geom_point(aes(x = Host_time,
      #              # y = mean_dist,
      #               y = Bact_dist,
      #               color = Host2),
      #             shape = "\u25D3", size = 3, stroke = 3
      #             ) +
      facet_wrap(~Species1, scales = "free_x", ncol = 2) +
      geom_smooth(aes(x = Host_time,
                    # y = mean_dist), color = "#000000", method = "lm", se = F) +
                    y = Bact_dist), color = "#000000", method = "lm", se = F) +
      annotate("text", x = get_stat_label(df_plot)[["x"]],
                    y = get_stat_label(df_plot)[["y"]],
                    size = 5,
        label = paste0("r2 = ", round(get_stat_label(df_plot)[["r"]], 4), ", p = ", round(get_stat_label(df_plot)[["r"]], 3), ", Slope = ", round(get_stat_label(df_plot)[["slope"]], 5))) +
      # labs(x = "Host divergence", y = "Bacteria divergence", title = paste0(genus)) +
      scale_color_manual(values = host_order_color_dark) +
      scale_fill_manual(values = host_order_color_dark) +
      scale_x_continuous(labels=unit_format(unit = "Mya", scale = 1)) +
      make_theme(theme_name=theme_few(), leg_pos="none",
                # guide_nrow = 6,
                setFill = F, setCol = F,
                x_angle=45 ,x_vj=1, x_hj=1, x_size=10,
                y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
    p_list[[genus]] <- p
  }
pdf(paste0("results/figures/12-phylo_distances/bact_host_dists_mean_species.pdf"), width = 15, height = 10)
# pdf(paste0("results/figures/12-phylo_distances/bact_host_dists_all_species.pdf"), width = 15, height = 10)
print(marrangeGrob(p_list, nrow = 2, ncol = 2))
dev.off()

# genus <- "Snodgrassella"
for (genus in genera_clean) {
  if (!file.exists(paste0("results/figures/12-phylo_distances/", genus,"/bact_host_dists.tsv"))) {
    next
  }
  
  # for a given species as species1 and species2, plot  Host_dist in x axis and Bact_dist in y axis
  df_phy_dist_same_spec <- df_phy_dist %>% filter(Species1 == Species2)
  p_list <- list()
  for (species in unique(df_phy_dist_same_spec$Species1)) {
    df_plot <- df_phy_dist_same_spec %>% filter(Species1 == species)
    if (length(unique(df_plot$Host1)) == 1) {
      next
    }
    p <- ggplot(df_plot) +
      geom_point(aes(x = Host_time,
                    y = Bact_dist,
                    fill = Host2,
                    color = Host1),
                  shape = 21, size = 3, stroke = 3
                  ) +
      # geom_point(aes(x = Host_time,
      #               y = Bact_dist,
      #               color = Host1),
      #             shape = "\u25D2", size = 3, stroke = 3
      #             ) +
      # geom_point(aes(x = Host_time,
      #               y = Bact_dist,
      #               color = Host2),
      #             shape = "\u25D3", size = 3, stroke = 3
      #             ) +
      geom_smooth(aes(x = Host_time,
                    y = Bact_dist), color = "#000000", method = "lm", se = F) +
      annotate("text", x = get_stat_label(df_plot)[["x"]],
                    y = get_stat_label(df_plot)[["y"]],
                    size = 5,
        label = paste0("r2 = ", round(get_stat_label(df_plot)[["r"]], 4), ", p = ", round(get_stat_label(df_plot)[["r"]], 3), ", Slope = ", round(get_stat_label(df_plot)[["slope"]], 5))) +
      labs(x = "Host divergence", y = "Bacteria divergence", title = paste0(species)) +
      scale_color_manual(values = host_order_color_dark) +
      scale_fill_manual(values = host_order_color_dark) +
      scale_x_continuous(labels=unit_format(unit = "Mya", scale = 1)) +
      make_theme(theme_name=theme_few(), leg_pos="none",
                # guide_nrow = 6,
                setFill = F, setCol = F,
                x_angle=45 ,x_vj=1, x_hj=1, x_size=10,
                y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
    p_list[[species]] <- p
  }
  pdf(paste0("results/figures/12-phylo_distances/", genus,"/bact_host_dists_same_species.pdf"), width = 15, height = 10)
  print(marrangeGrob(p_list, nrow = 2, ncol = 2))
  dev.off()
}
# for (genus in genera_clean) {
#   if (!file.exists(paste0("results/figures/12-phylo_distances/", genus,"/bact_host_dists.tsv"))) {
#     next
#   }
#   df_phy_dist <- read.csv(paste0("results/figures/12-phylo_distances/", genus,"/bact_host_dists.tsv"), sep = "\t") %>%
#                   mutate(spec_comb = ifelse(Species1 == Species2, "Same", "Diff")) %>%
#                   filter(Host1 != "Other") %>%
#                   filter(Host2 != "Other")
#   # for a given species as species1 and species2, plot  Host_dist in x axis and Bact_dist in y axis
#   df_phy_dist_same_spec <- df_phy_dist %>% filter(Species1 == Species2)
#   p_list <- list()
#   for (species in unique(df_phy_dist_same_spec$Species1)) {
#     df_plot <- df_phy_dist_same_spec %>% filter(Species1 == species)
#     if (length(unique(df_plot$Host1)) == 1) {
#       next
#     }
#     p <- ggplot(df_plot) +
#       geom_point(aes(x = Host_time,
#                     y = Bact_dist,
#                     fill = Host2,
#                     color = Host1),
#                   shape = 21, size = 3, stroke = 3
#                   ) +
#       # geom_point(aes(x = Host_time,
#       #               y = Bact_dist,
#       #               color = Host1),
#       #             shape = "\u25D2", size = 3, stroke = 3
#       #             ) +
#       # geom_point(aes(x = Host_time,
#       #               y = Bact_dist,
#       #               color = Host2),
#       #             shape = "\u25D3", size = 3, stroke = 3
#       #             ) +
#       geom_smooth(aes(x = Host_time,
#                     y = Bact_dist), color = "#000000", method = "lm", se = F) +
#       annotate("text", x = get_stat_label(df_plot)[["x"]],
#                     y = get_stat_label(df_plot)[["y"]],
#                     size = 5,
#         label = paste0("r2 = ", round(get_stat_label(df_plot)[["r"]], 4), ", p = ", round(get_stat_label(df_plot)[["r"]], 3), ", Slope = ", round(get_stat_label(df_plot)[["slope"]], 5))) +
#       labs(x = "Host divergence", y = "Bacteria divergence", title = paste0(species)) +
#       scale_color_manual(values = host_order_color_dark) +
#       scale_fill_manual(values = host_order_color_dark) +
#       scale_x_continuous(labels=unit_format(unit = "Mya", scale = 1)) +
#       make_theme(theme_name=theme_few(), leg_pos="none",
#                 # guide_nrow = 6,
#                 setFill = F, setCol = F,
#                 x_angle=45 ,x_vj=1, x_hj=1, x_size=10,
#                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
#     p_list[[species]] <- p
#   }
#   pdf(paste0("results/figures/12-phylo_distances/", genus,"/bact_host_dists_same_species.pdf"), width = 15, height = 10)
#   print(marrangeGrob(p_list, nrow = 2, ncol = 2))
#   dev.off()
# }
  
```

```{r}
# abundance_df_matrix_rel
# abundance_df_matrix_pa
# select gilli frisch
species_names_df <- df_plot %>% select(MAG_species_name_final, magOTU) %>% unique
ab_df_pa_gf <- abundance_df_matrix_pa %>% as.data.frame %>%
                select(matches("g__Gilli|g__Frisch")) %>%
                rownames_to_column("sample") %>%
                pivot_longer(!sample, names_to = "magOTU", values_to = "abundance") %>%
                left_join(species_names_df, by = "magOTU") %>%
                mutate(magOTU = MAG_species_name_final) %>%
                select(!MAG_species_name_final) %>%
                pivot_wider(id_cols = sample, names_from = magOTU, values_from = abundance) %>%
                column_to_rownames("sample") %>%
                as.matrix
ab_df_rel_gf <- abundance_df_matrix_rel %>% as.data.frame %>%
                select(matches("g__Gilli|g__Frisch")) %>%
                rownames_to_column("sample") %>%
                pivot_longer(!sample, names_to = "magOTU", values_to = "abundance") %>%
                left_join(species_names_df, by = "magOTU") %>%
                mutate(magOTU = MAG_species_name_final) %>%
                select(!MAG_species_name_final) %>%
                pivot_wider(id_cols = sample, names_from = magOTU, values_from = abundance) %>%
                column_to_rownames("sample") %>%
                as.matrix

host_names <- rownames(ab_df_pa_gf) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host)
col_split <- as.factor(host_names)
genus_name <- colnames(ab_df_pa_gf) %>% as.data.frame() %>% 
                mutate(Genus = ifelse(grepl("Gilli", .), "Gilliamella", NA)) %>% 
                mutate(Genus = ifelse(grepl("Frisch", .), "Frischella", Genus)) %>% 
                pull(Genus)
row_split <- as.factor(genus_name)
heatmap_obj = Heatmap(t(ab_df_pa_gf),
            col = c("white", "black"),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            # left_annotation = anno_host_row,
            column_names_gp = grid::gpar(fontsize = 10),
            row_names_gp = grid::gpar(fontsize = 10),
            heatmap_legend_param = list(title = "Number genes copies"),
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_columns = T,
            cluster_rows = T
            )
draw(heatmap_obj, merge_legend = TRUE)
df_pa_gf <- abundance_df_matrix_pa %>% as.data.frame %>%
                select(matches("g__Gilli|g__Frisch")) %>%
                rownames_to_column("sample") %>%
                pivot_longer(!sample, names_to = "magOTU", values_to = "abundance") %>%
                left_join(species_names_df, by = "magOTU") %>%
                mutate(magOTU = MAG_species_name_final) %>%
                select(!MAG_species_name_final)
df_rel_gf <- abundance_df_matrix_rel %>% as.data.frame %>%
                select(matches("g__Gilli|g__Frisch")) %>%
                rownames_to_column("sample") %>%
                pivot_longer(!sample, names_to = "magOTU", values_to = "abundance") %>%
                left_join(species_names_df, by = "magOTU") %>%
                mutate(magOTU = MAG_species_name_final) %>%
                select(!MAG_species_name_final)

gilli_list <- unique(df_pa_gf %>% filter(grepl("Gilli", magOTU)) %>% pull(magOTU))
# "Gilliamella sp019469185", "Gilliamella apicola_F"  
# "Gilliamella SGB-76_1", "Gilliamella sp019469205"
# "Gilliamella apicola_E", "Gilliamella apicola_I"  
# "Gilliamella apicola_J", "Gilliamella SGB-78_2"   
# "Gilliamella sp945271295", "Gilliamella apicola_L"  
# "Gilliamella SGB-83_1", "Gilliamella SGB-82_1"   
# "Gilliamella SGB-79_1", "Gilliamella apicola_N"  
# "Gilliamella apicola", "Gilliamella apicola_Q"
# "Gilliamella apis", "Gilliamella apis_A" 
frisch_list <- unique(df_pa_gf %>% filter(grepl("Frisch", magOTU)) %>% pull(magOTU))
# "Frischella japonica" "Frischella SGB-92_1" "Frischella SGB-94_1" "Frischella perrara"

list_p <- list()
for (magotu in c("Gilliamella apicola_Q", "Gilliamella apicola_N", "Gilliamella apicola",
                 "Gilliamella apis", "Gilliamella apis_A")) {
  print(magotu)
  p <- ggplot(df_rel_gf %>%
      mutate(Genus = ifelse(grepl("Gilli", magOTU), "Gilliamella", NA)) %>% 
      mutate(Genus = ifelse(grepl("Frisch", magOTU), "Frischella", Genus)) %>%
      filter(magOTU %in% c(magotu, "Frischella perrara")) %>%
      mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
      filter(host %in% c("Apis mellifera")),
      aes(x = magOTU, y = abundance, group = sample, color = host)) +
  geom_point() +
  geom_line() +
  labs(x = "", y = "Log(rel. abundance)") +
  scale_color_manual(values = host_order_color_dark) +
  scale_fill_manual(values = host_order_color_dark) +
  scale_y_continuous(trans = "pseudo_log") +
  make_theme(setCol = F, setFill = F, x_angle = 50, x_hj = 1, x_vj = 1, leg_pos = "none")
  list_p[[magotu]] <- p
} 
marrangeGrob(list_p, nrow = 3, ncol = 2)

list_p <- list()
for (magotu in c("Gilliamella apicola_E", "Gilliamella apicola_F", "Gilliamella apicola_I",
                 "Gilliamella sp019469205", "Gilliamella sp019469185")) {
  print(magotu)
  p <- ggplot(df_rel_gf %>%
      mutate(Genus = ifelse(grepl("Gilli", magOTU), "Gilliamella", NA)) %>% 
      mutate(Genus = ifelse(grepl("Frisch", magOTU), "Frischella", Genus)) %>%
      filter(magOTU %in% c(magotu, "Frischella japonica")) %>%
      mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
      filter(host %in% c("Apis cerana")),
      aes(x = magOTU, y = abundance, group = sample, color = host)) +
  geom_point() +
  geom_line() +
  labs(x = "", y = "Log(rel. abundance)") +
  scale_color_manual(values = host_order_color_dark) +
  scale_fill_manual(values = host_order_color_dark) +
  scale_y_continuous(trans = "pseudo_log") +
  make_theme(setCol = F, setFill = F, x_angle = 50, x_hj = 1, x_vj = 1, leg_pos = "none")
  list_p[[magotu]] <- p
} 
marrangeGrob(list_p, nrow = 3, ncol = 2)

list_p <- list()
for (magotu in c("Gilliamella apicola_E", "Gilliamella apicola_F", "Gilliamella apicola_I", "Gilliamella apicola_J",
                 "Gilliamella  SGB-79_1", "Gilliamella  SGB-82_1", "Gilliamella  SGB-78_1", "Gilliamella  SGB-83_1", "Gilliamella  SGB-76_1",
                 "Gilliamella apicola_L", "Gilliamella sp945271295",
                 "Gilliamella sp019469205", "Gilliamella sp019469185")) {
  print(magotu)
  p <- ggplot(df_rel_gf %>%
      mutate(Genus = ifelse(grepl("Gilli", magOTU), "Gilliamella", NA)) %>% 
      mutate(Genus = ifelse(grepl("Frisch", magOTU), "Frischella", Genus)) %>%
      filter(magOTU %in% c(magotu, "Frischella SGB-92_1")) %>%
      mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
      filter(host %in% c("Apis dorsata")),
      aes(x = magOTU, y = abundance, group = sample, color = host)) +
  geom_point() +
  geom_line() +
  labs(x = "", y = "Log(rel. abundance)") +
  scale_color_manual(values = host_order_color_dark) +
  scale_fill_manual(values = host_order_color_dark) +
  scale_y_continuous(trans = "pseudo_log") +
  make_theme(setCol = F, setFill = F, x_angle = 50, x_hj = 1, x_vj = 1, leg_pos = "none")
  list_p[[magotu]] <- p
} 
marrangeGrob(list_p, nrow = 4, ncol = 4)

list_p <- list()
for (magotu in c("Gilliamella apicola_E", "Gilliamella apicola_F", "Gilliamella apicola_I", "Gilliamella apicola_J",
                 "Gilliamella  SGB-79_1", "Gilliamella  SGB-82_1", "Gilliamella  SGB-78_1", "Gilliamella  SGB-83_1", "Gilliamella  SGB-76_1",
                 "Gilliamella apicola_L", "Gilliamella sp945271295",
                 "Gilliamella sp019469205", "Gilliamella sp019469185")) {
  print(magotu)
  p <- ggplot(df_rel_gf %>%
      mutate(Genus = ifelse(grepl("Gilli", magOTU), "Gilliamella", NA)) %>% 
      mutate(Genus = ifelse(grepl("Frisch", magOTU), "Frischella", Genus)) %>%
      filter(magOTU %in% c(magotu, "Frischella SGB-94_1")) %>%
      mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
      filter(host %in% c("Apis dorsata"))) +
  # geom_boxplot(aes(x = magOTU, y = abundance, fill = host), width = 0.5, alpha = 0.25) +
  geom_point(aes(x = magOTU, y = abundance, group = sample, color = host)) +
  geom_line(aes(x = magOTU, y = abundance, group = sample, color = host)) +
  labs(x = "", y = "Log(rel. abundance)") +
  scale_color_manual(values = host_order_color_dark) +
  scale_fill_manual(values = host_order_color_dark) +
  scale_y_continuous(trans = "pseudo_log") +
  make_theme(setCol = F, setFill = F, x_angle = 50, x_hj = 1, x_vj = 1, leg_pos = "none")
  list_p[[magotu]] <- p
} 
marrangeGrob(list_p, nrow = 4, ncol = 4)


ggplot(df_rel_gf %>%
      mutate(Genus = ifelse(grepl("Gilli", magOTU), "Gilliamella", NA)) %>% 
      mutate(Genus = ifelse(grepl("Frisch", magOTU), "Frischella", Genus)) %>%
      filter(magOTU %in% c("Frischella SGB-94_1", "Frischella SGB-92_1")) %>%
      mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
      filter(host %in% c("Apis dorsata"))) +
  geom_boxplot(aes(x = magOTU, y = abundance, fill = host), width = 0.1, alpha = 0.25) +
  geom_point(aes(x = magOTU, y = abundance, group = sample, color = host)) +
  geom_line(aes(x = magOTU, y = abundance, group = sample, color = host)) +
  labs(x = "", y = "Log (rel. abundance)") +
  scale_color_manual(values = host_order_color_dark) +
  scale_fill_manual(values = host_order_color_dark) +
  scale_y_continuous(trans = "pseudo_log") +
  make_theme(setFill = F, setCol = F, x_angle = 50, x_hj = 1, x_vj = 1, leg_pos = "none")


########################################################################################################################
# do this by host species
df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis mellifera")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Bifido", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis mellifera")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Lacto", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis mellifera")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Bombi", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis mellifera")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Gilli", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis mellifera")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Gilli", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis mellifera")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Snod", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")


########################################################################################################################

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis cerana")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Bifido", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis cerana")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Lacto", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis cerana")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Bombi", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis cerana")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Gilli", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis cerana")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Gilli", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis cerana")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Snod", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")


########################################################################################################################

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis dorsata")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Bifido", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis dorsata")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Lacto", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis dorsata")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Bombi", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis dorsata")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Gilli", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis dorsata")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Gilli|Frisch", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis dorsata")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Snod", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis dorsata")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Dysg", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")

df_mat_rel <- abundance_df_matrix_rel %>%
            as.data.frame %>%
            rownames_to_column("sample") %>%
            mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
            filter(host %in% c("Apis dorsata")) %>%
            select(!host) %>%
            select_if(function(x) !all(x == 0)) %>%
            column_to_rownames("sample") %>%
            # remove columns with only 0s
            t %>%
            as.data.frame %>%
            rownames_to_column("magOTU") %>%
            left_join(species_names_df, by = "magOTU") %>%l
            mutate(magOTU = MAG_species_name_final) %>%
            select(!MAG_species_name_final) %>%
            filter(!is.na(magOTU)) %>%
            filter(grepl("Frisch", magOTU)) %>%
            select_if(function(x)  !all(x == 0)) %>%
            column_to_rownames("magOTU") %>%
              as.matrix %>%
              t
cor_data <- cor(df_mat_rel)
corrplot(cor_data, method = 'color', tl.srt = 45, tl.col = "#000000", tl.cex = 0.75, order = "AOE", type = "lower")
```